// sandpack-vue3@3.1.11/dist/unstyled.mjs
var zn = Object.defineProperty, Jn = Object.defineProperties;
var Yn = Object.getOwnPropertyDescriptors;
var Ft = Object.getOwnPropertySymbols;
var Zn = Object.prototype.hasOwnProperty, Wn = Object.prototype.propertyIsEnumerable;
var jt = (e, n, t) => n in e ? zn(e, n, { enumerable: !0, configurable: !0, writable: !0, value: t }) : e[n] = t, x = (e, n) => {
  for (var t in n || (n = {}))
    Zn.call(n, t) && jt(e, t, n[t]);
  if (Ft)
    for (var t of Ft(n))
      Wn.call(n, t) && jt(e, t, n[t]);
  return e;
}, O = (e, n) => Jn(e, Yn(n));
var ke = (e, n, t) => new Promise((o, l) => {
  var c = (f) => {
    try {
      a(t.next(f));
    } catch (y) {
      l(y);
    }
  }, d = (f) => {
    try {
      a(t.throw(f));
    } catch (y) {
      l(y);
    }
  }, a = (f) => f.done ? o(f.value) : Promise.resolve(f.value).then(c, d);
  a((t = t.apply(e, n)).next());
});
import { watch as H, getCurrentScope as Gn, onScopeDispose as Kn, unref as Xn, defineComponent as A, provide as Je, createVNode as s, Fragment as Q, ref as T, reactive as pt, toRaw as V, onMounted as xe, nextTick as Ie, computed as k, onBeforeMount as Qn, onBeforeUnmount as ae, onUnmounted as se, inject as es, createTextVNode as w, mergeProps as me, isVNode as ts, watchEffect as Rt } from "vue";
import { dequal as ns } from "dequal";
import "@stitches/core";
import { normalizePath as Ae, addPackageJSONIfNeeded as ss, loadSandpackClient as is, extractErrorDetails as os } from "@codesandbox/sandpack-client";
import ls from "lz-string";
import as from "ansi-to-vue3";
import rs from "clean-set";
import { closeBrackets as cs, closeBracketsKeymap as us } from "@codemirror/autocomplete";
import { HighlightStyle as ds, syntaxHighlighting as fs, bracketMatching as ps } from "@codemirror/language";
import { EditorView as He, ViewPlugin as pn, Decoration as be, keymap as It, highlightSpecialChars as ms, highlightActiveLine as vs, lineNumbers as hs } from "@codemirror/view";
import { css as ys } from "@codemirror/lang-css";
import { html as gs } from "@codemirror/lang-html";
import { javascript as Dt } from "@codemirror/lang-javascript";
import { tags as _, highlightTree as bs } from "@lezer/highlight";
import { StateEffect as Bt, EditorState as Cs, EditorSelection as ws, Annotation as Pt } from "@codemirror/state";
import { history as Ss, defaultKeymap as xs, historyKeymap as Es, indentMore as ks, indentLess as Ts, deleteGroupBackward as As } from "@codemirror/commands";
const Ls = {
  onCodeUpdate: {
    type: Function
  },
  code: {
    type: String,
    required: !0
  },
  filePath: {
    type: String
  },
  fileType: {
    type: String
  },
  showLineNumbers: {
    type: Boolean,
    default: !1
  },
  showInlineErrors: {
    type: Boolean,
    default: !1
  },
  wrapContent: {
    type: Boolean,
    default: !1
  },
  editorState: {
    type: String,
    default: "pristine"
  },
  readOnly: {
    type: Boolean,
    default: !1
  },
  showReadOnly: {
    type: Boolean,
    default: !0
  },
  decorators: {
    type: Array,
    default() {
      return [];
    }
  },
  initMode: {
    type: String,
    default: "lazy"
  },
  extensions: {
    type: Array,
    default() {
      return [];
    }
  },
  extensionsKeymap: {
    type: Array,
    default() {
      return [];
    }
  },
  additionalLanguages: {
    type: Array,
    default() {
      return [];
    }
  }
};
function Ns(e) {
  return Gn() ? (Kn(e), !0) : !1;
}
const Os = typeof window < "u", Ms = Os ? window : void 0;
function qt(e) {
  var t;
  const n = Xn(e);
  return (t = n == null ? void 0 : n.$el) != null ? t : n;
}
const et = () => {
};
function Fs(e, n, t = {}) {
  const { root: o, rootMargin: l = "0px", threshold: c = 0.1, window: d = Ms } = t, a = d && "IntersectionObserver" in d;
  let f = et;
  const y = a ? H(() => ({
    el: qt(e),
    root: qt(o)
  }), ({ el: h, root: u }) => {
    if (f(), !h)
      return;
    const C = new IntersectionObserver(n, {
      root: u,
      rootMargin: l,
      threshold: c
    });
    C.observe(h), f = () => {
      C.disconnect(), f = et;
    };
  }, {
    immediate: !0,
    flush: "post"
  }) : et, m = () => {
    f(), y();
  };
  return Ns(m), {
    isSupported: a,
    stop: m
  };
}
const Ue = (e) => {
  const n = e.lastIndexOf("/");
  return e.slice(n + 1);
}, js = (e, n) => {
  const t = (e[0] === "/" ? e.slice(1) : e).split("/"), o = [];
  if (t.length === 1)
    o.unshift(t[0]);
  else
    for (let l = 0; l < n.length; l++) {
      const c = n[l].split("/");
      for (let d = 1; d <= t.length; d++) {
        const a = t[t.length - d], f = c[c.length - d];
        if (o.length < d && o.unshift(a), a !== f)
          break;
      }
    }
  return o.length < t.length && o.unshift(".."), o.join("/");
}, $t = (e) => {
  let n = 0, t = 0, o = 0;
  if (e.startsWith("#")) {
    if (e.length < 7)
      return !0;
    n = parseInt(e.substr(1, 2), 16), t = parseInt(e.substr(3, 2), 16), o = parseInt(e.substr(5, 2), 16);
  } else {
    const c = e.replace("rgb(", "").replace("rgba(", "").replace(")", "").split(",");
    if (c.length < 3)
      return !0;
    n = parseInt(c[0], 10), t = parseInt(c[1], 10), o = parseInt(c[2], 10);
  }
  return (n * 299 + t * 587 + o * 114) / 1e3 < 128;
};
let Rs = 0;
const ze = () => (+(Date.now().toString(10).substr(0, 4) + Rs++)).toString(16), je = () => "", lt = () => je, mn = Object.getOwnPropertyDescriptors({
  toString: je
});
Object.defineProperties(je, mn);
Object.defineProperties(lt, mn);
const Is = {
  createTheme: je,
  css: lt,
  getCssText: je,
  keyframes: lt
}, Te = {
  colors: {
    surface1: "#ffffff",
    surface2: "#EFEFEF",
    surface3: "#F3F3F3",
    disabled: "#C5C5C5",
    base: "#323232",
    clickable: "#808080",
    hover: "#4D4D4D",
    accent: "#3973E0",
    error: "#EA3323",
    errorSurface: "#FCF1F0",
    warning: "#6A4516",
    warningSurface: "#FEF2C0"
  },
  syntax: {
    plain: "#151515",
    comment: {
      color: "#999",
      fontStyle: "italic"
    },
    keyword: "#7C5AE3",
    tag: "#0971F1",
    punctuation: "#3B3B3B",
    definition: "#85A600",
    property: "#3B3B3B",
    static: "#3B3B3B",
    string: "#2E6BD0"
  },
  font: {
    body: '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
    mono: '"Fira Mono", "DejaVu Sans Mono", Menlo, Consolas, "Liberation Mono", Monaco, "Lucida Console", monospace',
    size: "13px",
    lineHeight: "20px"
  }
}, at = {
  colors: {
    surface1: "#151515",
    surface2: "#252525",
    surface3: "#2F2F2F",
    disabled: "#4D4D4D",
    base: "#808080",
    clickable: "#999999",
    hover: "#C5C5C5",
    accent: "#E5E5E5",
    error: "#FFB4A6",
    errorSurface: "#690000",
    warning: "#E7C400",
    warningSurface: "#3A3000"
  },
  syntax: {
    plain: "#FFFFFF",
    comment: {
      color: "#757575",
      fontStyle: "italic"
    },
    keyword: "#77B7D7",
    tag: "#DFAB5C",
    punctuation: "#ffffff",
    definition: "#86D9CA",
    property: "#77B7D7",
    static: "#C64640",
    string: "#977CDC"
  },
  font: {
    body: '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
    mono: '"Fira Mono", "DejaVu Sans Mono", Menlo, Consolas, "Liberation Mono", Monaco, "Lucida Console", monospace',
    size: "13px",
    lineHeight: "20px"
  }
};
var fn;
const Ds = {
  light: Te,
  dark: at,
  auto: typeof window < "u" && (fn = window == null ? void 0 : window.matchMedia) != null && fn.call(window, "(prefers-color-scheme: dark)").matches ? at : Te
}, U = "sp", { createTheme: _t, css: $l, getCssText: _l, keyframes: vn } = Is, Bs = {
  space: new Array(11).fill(" ").reduce((e, n, t) => O(x({}, e), {
    [t + 1]: `${(t + 1) * 4}px`
  }), {}),
  border: {
    radius: "4px"
  },
  layout: {
    height: "300px",
    headerHeight: "40px"
  },
  transitions: {
    default: "150ms ease"
  },
  zIndices: {
    base: "1",
    overlay: "2",
    top: "3"
  }
}, Ht = (e) => {
  const t = Object.entries(e.syntax).reduce((o, [l, c]) => {
    let d = {
      [`color-${l}`]: c
    };
    return typeof c == "object" && (d = Object.entries(c).reduce((a, [f, y]) => O(x({}, a), {
      [`${f}-${l}`]: y
    }), {})), x(x({}, o), d);
  }, {});
  return O(x({}, Bs), {
    colors: e.colors,
    font: e.font,
    syntax: t
  });
}, Vt = (e) => {
  var y, m, h, u, C;
  const n = "default";
  if (e = e || "light", typeof e == "string") {
    const r = Ds[e];
    if (!r)
      throw new Error(`[sandpack-vue3]: invalid theme '${e}' provided.`);
    return {
      theme: r,
      id: e,
      mode: $t(r.colors.surface1) ? "dark" : "light"
    };
  }
  const t = $t((m = (y = e == null ? void 0 : e.colors) == null ? void 0 : y.surface1) != null ? m : Te.colors.surface1) ? "dark" : "light", o = t === "dark" ? at : Te, l = x(x({}, o.colors), (h = e == null ? void 0 : e.colors) != null ? h : {}), c = x(x({}, o.syntax), (u = e == null ? void 0 : e.syntax) != null ? u : {}), d = x(x({}, o.font), (C = e == null ? void 0 : e.font) != null ? C : {}), a = {
    colors: l,
    syntax: c,
    font: d
  }, f = e ? Ps(JSON.stringify(a)) : n;
  return {
    theme: a,
    id: `sp-${f}`,
    mode: t
  };
}, Ps = (e) => {
  let n = 0;
  for (let t = 0; t < e.length; n &= n)
    n = 31 * n + e.charCodeAt(t++);
  return Math.abs(n);
}, g = () => "";
g.toString = g;
const rt = (e, { line: n, column: t }) => e.line(n).from + (t != null ? t : 0) - 1, qs = () => He.theme({
  "&": {
    backgroundColor: `var(--${U}-colors-surface1)`,
    color: `var(--${U}-syntax-color-plain)`,
    height: "100%"
  },
  ".cm-matchingBracket, .cm-nonmatchingBracket, &.cm-focused .cm-matchingBracket, &.cm-focused .cm-nonmatchingBracket": {
    color: "inherit",
    backgroundColor: "rgba(128,128,128,.25)",
    backgroundBlendMode: "difference"
  },
  "&.cm-editor.cm-focused": {
    outline: "none"
  },
  ".cm-activeLine": {
    backgroundColor: `var(--${U}-colors-surface3)`,
    borderRadius: `var(--${U}-border-radius)`
  },
  ".cm-errorLine": {
    backgroundColor: `var(--${U}-colors-errorSurface)`,
    borderRadius: `var(--${U}-border-radius)`
  },
  ".cm-content": {
    caretColor: `var(--${U}-colors-accent)`,
    padding: `0 var(--${U}-space-4)`
  },
  ".cm-scroller": {
    fontFamily: `var(--${U}-font-mono)`,
    lineHeight: `var(--${U}-font-lineHeight)`
  },
  ".cm-gutters": {
    backgroundColor: `var(--${U}-colors-surface1)`,
    color: `var(--${U}-colors-disabled)`,
    border: "none",
    paddingLeft: `var(--${U}-space-1)`
  },
  ".cm-gutter.cm-lineNumbers": {
    fontSize: ".6em"
  },
  ".cm-lineNumbers .cm-gutterElement": {
    lineHeight: `var(--${U}-font-lineHeight)`,
    minWidth: `var(--${U}-space-5)`
  },
  ".cm-content .cm-line": {
    paddingLeft: `var(--${U}-space-1)`
  },
  ".cm-content.cm-readonly .cm-line": {
    paddingLeft: 0
  }
}), fe = (e) => `${U}-syntax-${e}`, $s = (e) => ds.define([{
  tag: _.link,
  textDecoration: "underline"
}, {
  tag: _.emphasis,
  fontStyle: "italic"
}, {
  tag: _.strong,
  fontWeight: "bold"
}, {
  tag: _.keyword,
  class: fe("keyword")
}, {
  tag: [_.atom, _.number, _.bool],
  class: fe("static")
}, {
  tag: _.variableName,
  class: fe("plain")
}, {
  tag: _.function(_.variableName),
  class: fe("definition")
}, {
  tag: _.standard(_.tagName),
  class: fe("tag")
}, {
  tag: [_.function(_.variableName), _.definition(_.function(_.variableName)), _.tagName],
  class: fe("definition")
}, {
  tag: _.propertyName,
  class: fe("property")
}, {
  tag: [_.literal, _.inserted],
  class: fe(e.syntax.string ? "string" : "static")
}, {
  tag: _.punctuation,
  class: fe("punctuation")
}, {
  tag: [_.comment, _.quote],
  class: fe("comment")
}]), _s = (e, n, t) => {
  if (!e && !n)
    return "javascript";
  let o = n;
  if (!o && e) {
    const l = e.lastIndexOf(".");
    o = e.slice(l + 1);
  }
  for (const l of t)
    if (o === l.name || l.extensions.includes(o || ""))
      return l.name;
  switch (o) {
    case "ts":
    case "tsx":
      return "typescript";
    case "html":
    case "svelte":
    case "vue":
    case "astro":
      return "html";
    case "css":
    case "less":
    case "scss":
      return "css";
    case "js":
    case "jsx":
    case "json":
    default:
      return "javascript";
  }
}, Hs = (e, n) => {
  const t = {
    javascript: Dt({
      jsx: !0,
      typescript: !1
    }),
    typescript: Dt({
      jsx: !0,
      typescript: !0
    }),
    html: gs(),
    css: ys()
  };
  for (const o of n)
    if (e === o.name)
      return o.language;
  return t[e];
};
function Vs(e) {
  return pn.fromClass(class {
    constructor(n) {
      this.decorations = this.getDecoration(n);
    }
    update(n) {
    }
    getDecoration(n) {
      if (!e)
        return be.none;
      const t = e.map((o) => {
        var a, f, y;
        const l = be.line({
          attributes: {
            class: (a = o.className) != null ? a : ""
          }
        }), c = be.mark({
          class: (f = o.className) != null ? f : "",
          attributes: (y = o.elementAttributes) != null ? y : void 0
        }), d = rt(n.state.doc, {
          line: o.line,
          column: o.startColumn
        }) + 1;
        if (o.startColumn && o.endColumn) {
          const m = rt(n.state.doc, {
            line: o.line,
            column: o.endColumn
          }) + 1;
          return c.range(d, m);
        }
        return l.range(d);
      });
      return be.set(t);
    }
  }, {
    decorations: (n) => n.decorations
  });
}
function Us() {
  return Js;
}
const zs = be.line({
  attributes: {
    class: "cm-errorLine"
  }
}), Js = pn.fromClass(class {
  constructor() {
    this.decorations = be.none;
  }
  update(e) {
    e.transactions.forEach((n) => {
      const t = n.annotation("show-error");
      if (t !== void 0) {
        const o = rt(e.view.state.doc, {
          line: t
        }) + 1;
        this.decorations = be.set([zs.range(o)]);
      } else
        n.annotation("remove-errors") && (this.decorations = be.none);
    });
  }
}, {
  decorations: (e) => e.decorations
}), Ut = (e, n) => {
  if (e.length !== n.length)
    return !1;
  let t = !0;
  for (let o = 0; o < e.length; o++)
    if (e[o] !== n[o]) {
      t = !1;
      break;
    }
  return t;
}, hn = Symbol("classNamesProvider"), Ys = A({
  name: "ClassNamesProvider",
  props: {
    classes: {
      type: Object,
      required: !1,
      default() {
        return {};
      }
    }
  },
  setup(e, { slots: n }) {
    return Je(hn, e.classes), () => s(Q, null, [n.default ? n.default() : null]);
  }
}), R = () => {
  const e = mt(hn, {});
  return function(t, o = []) {
    const l = `${U}-${t}`;
    return Zs(...o, l, e[l]);
  };
}, Zs = (...e) => e.filter(Boolean).join(" "), q = {
  "/styles.css": {
    code: `body {
  font-family: sans-serif;
  -webkit-font-smoothing: auto;
  -moz-font-smoothing: auto;
  -moz-osx-font-smoothing: grayscale;
  font-smoothing: auto;
  text-rendering: optimizeLegibility;
  font-smooth: always;
  -webkit-tap-highlight-color: transparent;
  -webkit-touch-callout: none;
}

h1 {
  font-size: 1.5rem;
}`
  }
}, Ws = {
  files: {
    "/src/styles.css": q["/styles.css"],
    "/src/pages/index.astro": {
      code: `---
import "../styles.css";
const data = "world";
---

<h1>Hello {data}</h1>

<style>
  h1 {
    font-size: 1.5rem;
  }
</style>`
    },
    "/src/env.d.ts": {
      code: '/// <reference types="astro/client" />'
    },
    ".env": {
      code: 'ASTRO_TELEMETRY_DISABLED="1"'
    },
    "/astro.config.mjs": {
      code: `import { defineConfig } from 'astro/config';

// https://astro.build/config
export default defineConfig({});`
    },
    "/tsconfig.json": {
      code: `{
  "extends": "astro/tsconfigs/base"
}`
    },
    "/package.json": {
      code: JSON.stringify({
        dependencies: {
          astro: "^1.9.2",
          "esbuild-wasm": "^0.17.19"
        },
        scripts: {
          dev: "astro dev",
          start: "astro dev",
          build: "astro build",
          preview: "astro preview",
          astro: "astro"
        }
      }, null, 2)
    }
  },
  main: "/src/pages/index.astro",
  environment: "node"
}, Gs = {
  files: O(x({}, q), {
    "/styles.css": q["/styles.css"],
    "/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/vite.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite + Lit + TS</title>
    <link rel="stylesheet" href="/styles.css" />
    <script type="module" src="/main.ts"><\/script>
  </head>
  <body>
    <my-element />
  </body>
</html>
`
    },
    "/main.ts": {
      code: `import { LitElement, html } from 'lit'
import { customElement, property } from 'lit/decorators.js'

@customElement('my-element')
export class MyElement extends LitElement {
  @property()
  msg = 'Hello world'

  render() {
    return html\`<h1>\${this.msg}</h1>\`
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'my-element': MyElement
  }
}
`
    },
    "/vite-env.d.ts": {
      code: '/// <reference types="vite/client" />'
    },
    "/tsconfig.json": {
      code: `{
  "compilerOptions": {
    "module": "ESNext",
    "lib": ["ES2020", "DOM", "DOM.Iterable"],
    "declaration": true,
    "emitDeclarationOnly": true,
    "outDir": "./types",
    "strict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "moduleResolution": "Node",
    "isolatedModules": true,
    "allowSyntheticDefaultImports": true,
    "experimentalDecorators": true,
    "forceConsistentCasingInFileNames": true,
    "useDefineForClassFields": false,
    "skipLibCheck": true
  },
  "references": [{ "path": "./tsconfig.node.json" }]
}
`
    },
    "/tsconfig.node.json": {
      code: `{
  "compilerOptions": {
    "composite": true,
    "module": "ESNext",
    "moduleResolution": "Node",
    "allowSyntheticDefaultImports": true
  },
  "include": ["vite.config.ts"]
}
`
    },
    "/vite.config.ts": {
      code: `import { defineConfig } from 'vite'

// https://vitejs.dev/config/
export default defineConfig({
  build: {
    lib: {
      entry: '/main.ts',
      formats: ['es'],
    },
    rollupOptions: {
      external: /^lit/,
    },
  },
})
`
    },
    "/package.json": {
      code: `{
  "name": "vite-lit",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "tsc && vite build"
  },
  "dependencies": {
    "lit": "^2.7.4"
  },
  "devDependencies": {
    "typescript": "^5.0.4",
    "vite": "4.2.2",
    "esbuild-wasm": "^0.17.19"
  }
}
`
    }
  }),
  entry: "/main.ts",
  main: "/main.ts",
  environment: "node"
}, Ks = {
  files: O(x({}, q), {
    "/pages/_app.js": {
      code: `import '../styles.css'

export default function MyApp({ Component, pageProps }) {
  return <Component {...pageProps} />
}`
    },
    "/pages/index.js": {
      code: `export default function Home({ data }) {
  return (
    <div>
      <h1>Hello {data}</h1>
    </div>
  );
}

export function getServerSideProps() {
  return {
    props: { data: "world" },
  }
}
`
    },
    "/next.config.js": {
      code: `/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  swcMinify: true,
}

module.exports = nextConfig
`
    },
    "/package.json": {
      code: JSON.stringify({
        name: "my-app",
        version: "0.1.0",
        private: !0,
        scripts: {
          dev: "NEXT_TELEMETRY_DISABLED=1 next dev",
          build: "next build",
          start: "next start",
          lint: "next lint"
        },
        dependencies: {
          next: "12.1.6",
          react: "18.2.0",
          "react-dom": "18.2.0",
          "@next/swc-wasm-nodejs": "12.1.6"
        }
      }, null, 2)
    }
  }),
  main: "/pages/index.js",
  environment: "node"
}, Xs = {
  files: {
    "/index.js": {
      code: `const http = require('http');

const hostname = '127.0.0.1';
const port = 3000;

const server = http.createServer((req, res) => {
  res.statusCode = 200;
  res.setHeader('Content-Type', 'text/html');
  res.end('Hello world');
});

server.listen(port, hostname, () => {
  console.log(\`Server running at http://\${hostname}:\${port}/\`);
});`
    },
    "/package.json": {
      code: JSON.stringify({
        dependencies: {},
        scripts: {
          start: "node index.js"
        },
        main: "index.js"
      }, null, 2)
    }
  },
  main: "/index.js",
  environment: "node"
}, Qs = {
  files: O(x({}, q), {
    "/App.tsx": {
      code: `import { Component } from "solid-js";

const App: Component = () => {
  return <h1>Hello world</h1>
};

export default App;`
    },
    "/index.tsx": {
      code: `import { render } from "solid-js/web";
import App from "./App";

import "./styles.css";

render(() => <App />, document.getElementById("app"));`
    },
    "/index.html": {
      code: `<html>
<head>
  <title>Parcel Sandbox</title>
  <meta charset="UTF-8" />
</head>
<body>
  <div id="app"></div>
  <script type="module" src="/index.tsx"><\/script>
</body>
</html>`
    },
    "/vite.config.ts": {
      code: `/// <reference types="vite/client" />

import { defineConfig } from 'vite';
import solidPlugin from 'vite-plugin-solid';

export default defineConfig({
  plugins: [solidPlugin()],
  build: {
    target: 'esnext',
  },
  resolve: {
    conditions: ['development', 'browser'],
  },
});`
    },
    "/tsconfig.json": {
      code: `{
  "compilerOptions": {
    "strict": true,
    "target": "ESNext",
    "module": "ESNext",
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "jsx": "preserve",
    "jsxImportSource": "solid-js",
    "types": ["vite/client", "@testing-library/jest-dom"]
  }
}`
    },
    "/package.json": {
      code: JSON.stringify({
        type: "module",
        scripts: {
          dev: "vite",
          build: "tsc && vite build",
          serve: "vite preview"
        },
        dependencies: {
          "solid-js": "^1.7.5"
        },
        devDependencies: {
          vite: "4.2.2",
          typescript: "^5.0.4",
          "vite-plugin-solid": "^2.7.0",
          "esbuild-wasm": "^0.17.19"
        },
        main: "/index.tsx"
      }, null, 2)
    }
  }),
  main: "/App.tsx",
  environment: "node"
}, ei = {
  files: O(x({}, q), {
    "/index.js": {
      code: `import "./styles.css";

document.getElementById("app").innerHTML = \`
<h1>Hello world</h1>
\`;
`
    },
    "/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
  </head>
  <body>
    <div id="app"></div>
    <script type="module" src="/index.js"><\/script>
  </body>
</html>
`
    },
    "/package.json": {
      code: JSON.stringify({
        scripts: {
          dev: "vite",
          build: "vite build",
          preview: "vite preview"
        },
        devDependencies: {
          vite: "4.2.2",
          "esbuild-wasm": "^0.17.19"
        }
      }, null, 2)
    }
  }),
  main: "/index.js",
  environment: "node"
}, ti = {
  files: O(x({}, q), {
    "/App.jsx": {
      code: `export default function App() {
  const data = "world"

  return <h1>Hello {data}</h1>
}
`
    },
    "/index.jsx": {
      code: `import { StrictMode } from "react";
import { createRoot } from "react-dom/client";
import "./styles.css";

import App from "./App";

const root = createRoot(document.getElementById("root"));
root.render(
  <StrictMode>
    <App />
  </StrictMode>
);`
    },
    "/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/index.jsx"><\/script>
  </body>
</html>
`
    },
    "/package.json": {
      code: JSON.stringify({
        scripts: {
          dev: "vite",
          build: "vite build",
          preview: "vite preview"
        },
        dependencies: {
          react: "^18.2.0",
          "react-dom": "^18.2.0"
        },
        devDependencies: {
          "@vitejs/plugin-react": "^3.1.0",
          vite: "4.2.2",
          "esbuild-wasm": "^0.17.19"
        }
      }, null, 2)
    },
    "/vite.config.js": {
      code: `import { defineConfig } from "vite";
import react from "@vitejs/plugin-react";
// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react()],
});
`
    }
  }),
  main: "/App.jsx",
  environment: "node"
}, ni = {
  files: O(x({}, q), {
    "/App.tsx": {
      code: `export default function App() {
  const data: string = "world"

  return <h1>Hello {data}</h1>
}
`
    },
    "/index.tsx": {
      code: `import { StrictMode } from "react";
import { createRoot } from "react-dom/client";
import "./styles.css";

import App from "./App";
import React from "react";

const root = createRoot(document.getElementById("root") as HTMLElement);
root.render(
  <StrictMode>
    <App />
  </StrictMode>
);
`
    },
    "/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/index.tsx"><\/script>
  </body>
</html>
`
    },
    "/tsconfig.json": {
      code: JSON.stringify({
        compilerOptions: {
          target: "ESNext",
          useDefineForClassFields: !0,
          lib: ["DOM", "DOM.Iterable", "ESNext"],
          allowJs: !1,
          skipLibCheck: !0,
          esModuleInterop: !1,
          allowSyntheticDefaultImports: !0,
          strict: !0,
          forceConsistentCasingInFileNames: !0,
          module: "ESNext",
          moduleResolution: "Node",
          resolveJsonModule: !0,
          isolatedModules: !0,
          noEmit: !0,
          jsx: "react-jsx"
        },
        references: [{
          path: "./tsconfig.node.json"
        }]
      }, null, 2)
    },
    "/tsconfig.node.json": {
      code: JSON.stringify({
        compilerOptions: {
          composite: !0,
          module: "ESNext",
          moduleResolution: "Node",
          allowSyntheticDefaultImports: !0
        },
        include: ["vite.config.ts"]
      }, null, 2)
    },
    "/package.json": {
      code: JSON.stringify({
        scripts: {
          dev: "vite",
          build: "tsc && vite build",
          preview: "vite preview"
        },
        dependencies: {
          react: "^18.2.0",
          "react-dom": "^18.2.0"
        },
        devDependencies: {
          "@types/react": "^18.0.28",
          "@types/react-dom": "^18.0.11",
          "@vitejs/plugin-react": "^3.1.0",
          "vite-plugin-checker": "^0.5.6",
          typescript: "^5.0.4",
          vite: "4.2.2",
          "esbuild-wasm": "^0.17.19"
        }
      }, null, 2)
    },
    "/vite-env.d.ts": {
      code: '/// <reference types="vite/client" />'
    },
    "/vite.config.ts": {
      code: `import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    react(),
  ],
})
`
    }
  }),
  main: "/App.tsx",
  environment: "node"
}, si = {
  files: O(x({}, q), {
    "/App.jsx": {
      code: `export default function App() {
  const data = "world"

  return <h1>Hello {data}</h1>
}
`
    },
    "/index.jsx": {
      code: `import { render } from "preact";
import "./styles.css";

import App from "./App";

const root = document.getElementById("root");
render(<App />, root);
`
    },
    "/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/index.jsx"><\/script>
  </body>
</html>
`
    },
    "/package.json": {
      code: JSON.stringify({
        scripts: {
          dev: "vite",
          build: "vite build",
          preview: "vite preview"
        },
        dependencies: {
          preact: "^10.16.0"
        },
        devDependencies: {
          "@preact/preset-vite": "^2.5.0",
          vite: "4.2.2",
          "esbuild-wasm": "0.17.12"
        }
      })
    },
    "/vite.config.js": {
      code: `import { defineConfig } from "vite";
import preact from '@preact/preset-vite'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [preact()],
});
`
    }
  }),
  main: "/App.jsx",
  environment: "node"
}, ii = {
  files: O(x({}, q), {
    "/App.tsx": {
      code: `export default function App() {
  const data: string = "world"

  return <h1>Hello {data}</h1>
}
`
    },
    "/index.tsx": {
      code: `import { render } from "preact";
import "./styles.css";

import App from "./App";

const root = document.getElementById("root") as HTMLElement;
render(<App />, root);
`
    },
    "/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/index.tsx"><\/script>
  </body>
</html>
`
    },
    "/tsconfig.json": {
      code: JSON.stringify({
        compilerOptions: {
          target: "ESNext",
          useDefineForClassFields: !0,
          lib: ["DOM", "DOM.Iterable", "ESNext"],
          allowJs: !1,
          skipLibCheck: !0,
          esModuleInterop: !1,
          allowSyntheticDefaultImports: !0,
          strict: !0,
          forceConsistentCasingInFileNames: !0,
          module: "ESNext",
          moduleResolution: "Node",
          resolveJsonModule: !0,
          isolatedModules: !0,
          noEmit: !0,
          jsx: "react-jsx",
          jsxImportSource: "preact"
        },
        include: ["src"],
        references: [{
          path: "./tsconfig.node.json"
        }]
      }, null, 2)
    },
    "/tsconfig.node.json": {
      code: JSON.stringify({
        compilerOptions: {
          composite: !0,
          module: "ESNext",
          moduleResolution: "Node",
          allowSyntheticDefaultImports: !0
        },
        include: ["vite.config.ts"]
      }, null, 2)
    },
    "/package.json": {
      code: JSON.stringify({
        scripts: {
          dev: "vite",
          build: "tsc && vite build",
          preview: "vite preview"
        },
        dependencies: {
          preact: "^10.16.0"
        },
        devDependencies: {
          "@preact/preset-vite": "^2.5.0",
          typescript: "^4.9.5",
          vite: "4.2.2",
          "esbuild-wasm": "^0.17.12"
        }
      }, null, 2)
    },
    "/vite-env.d.ts": {
      code: '/// <reference types="vite/client" />'
    },
    "/vite.config.ts": {
      code: `import { defineConfig } from 'vite'
import preact from '@preact/preset-vite'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [preact()],
})
`
    }
  }),
  main: "/App.tsx",
  environment: "node"
}, oi = {
  files: {
    "/src/styles.css": q["/styles.css"],
    "/src/App.svelte": {
      code: `<script>
const data = "world";
<\/script>

<h1>Hello {data}</h1>

<style>
h1 {
  font-size: 1.5rem;
}
</style>`
    },
    "/src/main.js": {
      code: `import App from './App.svelte'
import "./styles.css"

const app = new App({
  target: document.getElementById('app'),
})

export default app`
    },
    "/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
  </head>
  <body>
    <div id="app"></div>
    <script type="module" src="/src/main.js"><\/script>
  </body>
</html>
`
    },
    "svelte.config.js": {
      code: `import { vitePreprocess } from '@sveltejs/vite-plugin-svelte'

export default {
  // Consult https://svelte.dev/docs#compile-time-svelte-preprocess
  // for more information about preprocessors
  preprocess: vitePreprocess(),
}
`
    },
    "/vite.config.js": {
      code: `import { defineConfig } from 'vite'
import { svelte } from '@sveltejs/vite-plugin-svelte'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [svelte()],
})`
    },
    "/package.json": {
      code: JSON.stringify({
        type: "module",
        scripts: {
          dev: "vite"
        },
        dependencies: {
          svelte: "^3.59.1"
        },
        devDependencies: {
          "@sveltejs/vite-plugin-svelte": "2.1.1",
          vite: "4.2.2",
          "esbuild-wasm": "^0.17.19"
        }
      }, null, 2)
    }
  },
  main: "/src/App.svelte",
  environment: "node"
}, li = {
  files: {
    "/src/styles.css": q["/styles.css"],
    "/src/App.svelte": {
      code: `<script lang="ts">
const data: string = "world";
<\/script>

<h1>Hello {data}</h1>

<style>
h1 {
  font-size: 1.5rem;
}
</style>`
    },
    "/src/main.ts": {
      code: `import App from './App.svelte'
import "./styles.css"

const app = new App({
  target: document.getElementById('app'),
})

export default app`
    },
    "/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
  </head>
  <body>
    <div id="app"></div>
    <script type="module" src="/src/main.ts"><\/script>
  </body>
</html>
`
    },
    "/vite-env.d.ts": {
      code: `/// <reference types="svelte" />
/// <reference types="vite/client" />`
    },
    "svelte.config.js": {
      code: `import { vitePreprocess } from '@sveltejs/vite-plugin-svelte'

export default {
  // Consult https://svelte.dev/docs#compile-time-svelte-preprocess
  // for more information about preprocessors
  preprocess: vitePreprocess(),
}
`
    },
    "/vite.config.ts": {
      code: `import { defineConfig } from 'vite'
import { svelte } from '@sveltejs/vite-plugin-svelte'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [svelte()],
})`
    },
    "tsconfig.json": {
      code: JSON.stringify({
        extends: "@tsconfig/svelte/tsconfig.json",
        compilerOptions: {
          target: "ESNext",
          useDefineForClassFields: !0,
          module: "ESNext",
          resolveJsonModule: !0,
          allowJs: !0,
          checkJs: !0,
          isolatedModules: !0
        },
        include: ["src/**/*.d.ts", "src/**/*.ts", "src/**/*.js", "src/**/*.svelte"],
        references: [{
          path: "./tsconfig.node.json"
        }]
      }, null, 2)
    },
    "tsconfig.node.json": {
      code: JSON.stringify({
        compilerOptions: {
          composite: !0,
          module: "ESNext",
          moduleResolution: "Node"
        },
        include: ["vite.config.ts"]
      }, null, 2)
    },
    "/package.json": {
      code: JSON.stringify({
        type: "module",
        scripts: {
          dev: "vite"
        },
        devDependencies: {
          "@sveltejs/vite-plugin-svelte": "2.1.1",
          "@tsconfig/svelte": "^4.0.1",
          svelte: "^3.59.1",
          "svelte-check": "^3.3.2",
          typescript: "^5.0.4",
          tslib: "^2.5.0",
          vite: "4.2.2",
          "esbuild-wasm": "^0.17.19"
        }
      }, null, 2)
    }
  },
  main: "/src/App.svelte",
  environment: "node"
}, ai = {
  files: {
    "/src/styles.css": q["/styles.css"],
    "/src/App.vue": {
      code: `<script setup>
import { ref } from "vue";

const data = ref("world");
<\/script>

<template>
  <h1>Hello {{ data }}</h1>
</template>

<style>
h1 {
  font-size: 1.5rem;
}
</style>`
    },
    "/src/main.js": {
      code: `import { createApp } from 'vue'
import App from './App.vue'
import "./styles.css"

createApp(App).mount('#app')
`
    },
    "/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
  </head>
  <body>
    <div id="app"></div>
    <script type="module" src="/src/main.js"><\/script>
  </body>
</html>
`
    },
    "/vite.config.js": {
      code: `import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [vue()]
})
`
    },
    "/package.json": {
      code: JSON.stringify({
        scripts: {
          dev: "vite",
          build: "vite build",
          preview: "vite preview"
        },
        dependencies: {
          vue: "^3.3.2"
        },
        devDependencies: {
          "@vitejs/plugin-vue": "^4.2.3",
          vite: "4.2.2",
          "esbuild-wasm": "^0.17.19"
        }
      }, null, 2)
    }
  },
  main: "/src/App.vue",
  environment: "node"
}, ri = {
  files: {
    "/src/styles.css": q["/styles.css"],
    "/src/App.vue": {
      code: `<script setup lang="ts">
import { ref } from "vue";

const data = ref<string>("world");
<\/script>

<template>
  <h1>Hello {{ data }}</h1>
</template>

<style>
h1 {
  font-size: 1.5rem;
}
</style>`
    },
    "/src/main.ts": {
      code: `import { createApp } from 'vue'
import App from './App.vue'
import "./styles.css"

createApp(App).mount('#app')
`
    },
    "/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
  </head>
  <body>
    <div id="app"></div>
    <script type="module" src="/src/main.ts"><\/script>
  </body>
</html>
`
    },
    "/vite-env.d.ts": {
      code: '/// <reference types="vite/client" />'
    },
    "/vite.config.ts": {
      code: `import { defineConfig } from 'vite'
import vue from '@vitejs/plugin-vue'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [vue()]
})
`
    },
    "tsconfig.json": {
      code: JSON.stringify({
        compilerOptions: {
          target: "ESNext",
          useDefineForClassFields: !0,
          module: "ESNext",
          moduleResolution: "Node",
          strict: !0,
          jsx: "preserve",
          resolveJsonModule: !0,
          isolatedModules: !0,
          esModuleInterop: !0,
          lib: ["ESNext", "DOM"],
          skipLibCheck: !0,
          noEmit: !0
        },
        include: ["src/**/*.ts", "src/**/*.d.ts", "src/**/*.tsx", "src/**/*.vue"],
        references: [{
          path: "./tsconfig.node.json"
        }]
      }, null, 2)
    },
    "tsconfig.node.json": {
      code: JSON.stringify({
        compilerOptions: {
          composite: !0,
          module: "ESNext",
          moduleResolution: "Node",
          allowSyntheticDefaultImports: !0
        },
        include: ["vite.config.ts"]
      }, null, 2)
    },
    "/package.json": {
      code: JSON.stringify({
        scripts: {
          dev: "vite",
          build: "tsc && vite build",
          preview: "vite preview"
        },
        dependencies: {
          vue: "^3.3.2"
        },
        devDependencies: {
          "@vitejs/plugin-vue": "^4.2.3",
          vite: "4.2.2",
          "vue-tsc": "^1.6.5",
          typescript: "^5.0.4",
          "esbuild-wasm": "^0.17.19"
        }
      }, null, 2)
    }
  },
  main: "/src/App.vue",
  environment: "node"
}, ci = {
  files: {
    "/src/app/app.component.css": q["/styles.css"],
    "/src/app/app.component.html": {
      code: `<div>
<h1>{{ helloWorld }}</h1>
</div>
`
    },
    "/src/app/app.component.ts": {
      code: `import { Component } from "@angular/core";

@Component({
  selector: "app-root",
  templateUrl: "./app.component.html",
  styleUrls: ["./app.component.css"]
})
export class AppComponent {
  helloWorld = "Hello world";
}
`
    },
    "/src/app/app.module.ts": {
      code: `import { BrowserModule } from "@angular/platform-browser";
import { NgModule } from "@angular/core";

import { AppComponent } from "./app.component";

@NgModule({
  declarations: [AppComponent],
  imports: [BrowserModule],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule {}
`
    },
    "/src/index.html": {
      code: `<!doctype html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <title>Angular</title>
  <base href="/">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="icon" type="image/x-icon" href="favicon.ico">
</head>

<body>
   <app-root></app-root>
</body>

</html>
`
    },
    "/src/main.ts": {
      code: `import { enableProdMode } from "@angular/core";
import { platformBrowserDynamic } from "@angular/platform-browser-dynamic";

import { AppModule } from "./app/app.module";

platformBrowserDynamic()
  .bootstrapModule(AppModule)
  .catch(err => console.log(err));

`
    },
    "/src/polyfills.ts": {
      code: `import "core-js/proposals/reflect-metadata";
      import "zone.js/dist/zone";
`
    },
    "/package.json": {
      code: JSON.stringify({
        dependencies: {
          "@angular/core": "^11.2.0",
          "@angular/platform-browser": "^11.2.0",
          "@angular/platform-browser-dynamic": "^11.2.0",
          "@angular/common": "^11.2.0",
          "@angular/compiler": "^11.2.0",
          "zone.js": "0.11.3",
          "core-js": "3.8.3",
          rxjs: "6.6.3"
        },
        main: "/src/main.ts"
      }, null, 2)
    }
  },
  main: "/src/app/app.component.ts",
  environment: "angular-cli"
}, ui = {
  files: O(x({}, q), {
    "/App.js": {
      code: `export default function App() {
  return <h1>Hello world</h1>
}
`
    },
    "/index.js": {
      code: `import React, { StrictMode } from "react";
import { createRoot } from "react-dom/client";
import "./styles.css";

import App from "./App";

const root = createRoot(document.getElementById("root"));
root.render(
  <StrictMode>
    <App />
  </StrictMode>
);`
    },
    "/public/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>`
    },
    "/package.json": {
      code: JSON.stringify({
        dependencies: {
          react: "^18.0.0",
          "react-dom": "^18.0.0",
          "react-scripts": "^5.0.0"
        },
        main: "/index.js"
      }, null, 2)
    }
  }),
  main: "/App.js",
  environment: "create-react-app"
}, di = {
  files: O(x({}, q), {
    "tsconfig.json": {
      code: `{
  "include": [
    "./**/*"
  ],
  "compilerOptions": {
    "strict": true,
    "esModuleInterop": true,
    "lib": [ "dom", "es2015" ],
    "jsx": "react-jsx"
  }
}`
    },
    "/App.tsx": {
      code: `export default function App(): JSX.Element {
  return <h1>Hello world</h1>
}
`
    },
    "/index.tsx": {
      code: `import React, { StrictMode } from "react";
import { createRoot } from "react-dom/client";
import "./styles.css";

import App from "./App";

const root = createRoot(document.getElementById("root"));
root.render(
  <StrictMode>
    <App />
  </StrictMode>
);`
    },
    "/public/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>`
    },
    "/package.json": {
      code: JSON.stringify({
        dependencies: {
          react: "^18.0.0",
          "react-dom": "^18.0.0",
          "react-scripts": "^5.0.0"
        },
        devDependencies: {
          "@types/react": "^18.0.0",
          "@types/react-dom": "^18.0.0",
          typescript: "^4.0.0"
        },
        main: "/index.tsx"
      }, null, 2)
    }
  }),
  main: "/App.tsx",
  environment: "create-react-app"
}, fi = {
  files: O(x({}, q), {
    "/App.tsx": {
      code: `import { Component } from "solid-js";

const App: Component = () => {
  return <h1>Hello world</h1>
};

export default App;`
    },
    "/index.tsx": {
      code: `import { render } from "solid-js/web";
import App from "./App";

import "./styles.css";

render(() => <App />, document.getElementById("app"));`
    },
    "/index.html": {
      code: `<html>
<head>
  <title>Parcel Sandbox</title>
  <meta charset="UTF-8" />
</head>
<body>
  <div id="app"></div>
  <script src="src/index.tsx"><\/script>
</body>
</html>`
    },
    "/package.json": {
      code: JSON.stringify({
        dependencies: {
          "solid-js": "1.3.15"
        },
        main: "/index.tsx"
      }, null, 2)
    }
  }),
  main: "/App.tsx",
  environment: "solid"
}, pi = {
  files: O(x({}, q), {
    "/App.svelte": {
      code: `<style>
  h1 {
    font-size: 1.5rem;
  }
</style>

<script>
  let name = 'world';
<\/script>

<main>
  <h1>Hello {name}</h1>
</main>`
    },
    "/index.js": {
      code: `import App from "./App.svelte";
import "./styles.css";

const app = new App({
  target: document.body
});

export default app;
      `
    },
    "/public/index.html": {
      code: `<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf8" />
    <meta name="viewport" content="width=device-width" />

    <title>Svelte app</title>

    <link rel="stylesheet" href="public/bundle.css" />
  </head>

  <body>
    <script src="bundle.js"><\/script>
  </body>
</html>`
    },
    "/package.json": {
      code: JSON.stringify({
        dependencies: {
          svelte: "^3.0.0"
        },
        main: "/index.js"
      }, null, 2)
    }
  }),
  main: "/App.svelte",
  environment: "svelte"
}, mi = {
  files: {
    "tsconfig.json": {
      code: `{
  "include": [
    "./**/*"
  ],
  "compilerOptions": {
    "strict": true,
    "esModuleInterop": true,
    "lib": [ "dom", "es2015" ],
    "jsx": "react-jsx"
  }
}`
    },
    "/add.ts": {
      code: "export const add = (a: number, b: number): number => a + b;"
    },
    "/add.test.ts": {
      code: `import { add } from './add';

describe('add', () => {
  test('Commutative Law of Addition', () => {
    expect(add(1, 2)).toBe(add(2, 1));
  });
});`
    },
    "package.json": {
      code: JSON.stringify({
        dependencies: {},
        devDependencies: {
          typescript: "^4.0.0"
        },
        main: "/add.ts"
      }, null, 2)
    }
  },
  main: "/add.test.ts",
  environment: "parcel",
  mode: "tests"
}, vi = {
  files: O(x({}, q), {
    "/index.js": {
      code: `import "./styles.css";

document.getElementById("app").innerHTML = \`
<h1>Hello world</h1>
\`;
`
    },
    "/index.html": {
      code: `<!DOCTYPE html>
<html>

<head>
  <title>Parcel Sandbox</title>
  <meta charset="UTF-8" />
</head>

<body>
  <div id="app"></div>

  <script src="index.js">
  <\/script>
</body>

</html>`
    },
    "/package.json": {
      code: JSON.stringify({
        dependencies: {},
        main: "/index.js"
      }, null, 2)
    }
  }),
  main: "/index.js",
  environment: "parcel"
}, hi = {
  files: O(x({}, q), {
    "tsconfig.json": {
      code: `{
  "compilerOptions": {
    "strict": true,
    "module": "commonjs",
    "jsx": "preserve",
    "esModuleInterop": true,
    "sourceMap": true,
    "allowJs": true,
    "lib": [
      "es6",
      "dom"
    ],
    "rootDir": "src",
    "moduleResolution": "node"
  }
}`
    },
    "/index.ts": {
      code: `import "./styles.css";

document.getElementById("app").innerHTML = \`
<h1>Hello world</h1>
\`;
`
    },
    "/index.html": {
      code: `<!DOCTYPE html>
<html>

<head>
  <title>Parcel Sandbox</title>
  <meta charset="UTF-8" />
</head>

<body>
  <div id="app"></div>

  <script src="index.ts">
  <\/script>
</body>

</html>`
    },
    "/package.json": {
      code: JSON.stringify({
        dependencies: {},
        devDependencies: {
          typescript: "^4.0.0"
        },
        main: "/index.ts"
      }, null, 2)
    }
  }),
  main: "/index.ts",
  environment: "parcel"
}, zt = {
  files: {
    "/src/styles.css": q["/styles.css"],
    "/src/App.vue": {
      code: `<template>
  <h1>Hello {{ msg }}</h1>
</template>

<script setup>
import { ref } from 'vue';
const msg = ref('world');
<\/script>`
    },
    "/src/main.js": {
      code: `import { createApp } from 'vue'
import App from './App.vue'
import "./styles.css";

createApp(App).mount('#app')
`
    },
    "/public/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0" />
    <title>codesandbox</title>
  </head>
  <body>
    <noscript>
      <strong
        >We're sorry but codesandbox doesn't work properly without JavaScript
        enabled. Please enable it to continue.</strong
      >
    </noscript>
    <div id="app"></div>
    <!-- built files will be auto injected -->
  </body>
</html>
`
    },
    "/package.json": {
      code: JSON.stringify({
        name: "vue3",
        version: "0.1.0",
        private: !0,
        main: "/src/main.js",
        scripts: {
          serve: "vue-cli-service serve",
          build: "vue-cli-service build"
        },
        dependencies: {
          "core-js": "^3.30.2",
          vue: "^3.3.2"
        },
        devDependencies: {
          "@vue/cli-plugin-babel": "^5.0.8",
          "@vue/cli-service": "^5.0.8"
        }
      }, null, 2)
    }
  },
  main: "/src/App.vue",
  environment: "vue-cli"
}, Jt = {
  files: {
    "/src/styles.css": q["/styles.css"],
    "/src/App.vue": {
      code: `<template>
  <h1>Hello {{ msg }}</h1>
</template>

<script setup lang="ts">
import { ref } from 'vue';
const msg = ref<string>('world');
<\/script>`
    },
    "/src/main.ts": {
      code: `import { createApp } from 'vue'
import App from './App.vue'
import "./styles.css";

createApp(App).mount('#app')
`
    },
    "/src/shims-vue.d.ts": `/* eslint-disable */
declare module "*.vue" {
  import type { DefineComponent } from "vue";
  const component: DefineComponent<{}, {}, any>;
  export default component;
}`,
    "/public/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0" />
    <title>codesandbox</title>
  </head>
  <body>
    <noscript>
      <strong
        >We're sorry but codesandbox doesn't work properly without JavaScript
        enabled. Please enable it to continue.</strong
      >
    </noscript>
    <div id="app"></div>
    <!-- built files will be auto injected -->
  </body>
</html>
`
    },
    "/package.json": {
      code: JSON.stringify({
        name: "vue3-ts",
        version: "0.1.0",
        private: !0,
        main: "/src/main.ts",
        scripts: {
          serve: "vue-cli-service serve",
          build: "vue-cli-service build"
        },
        dependencies: {
          "core-js": "^3.30.2",
          vue: "^3.3.2"
        },
        devDependencies: {
          "@vue/cli-plugin-babel": "^5.0.8",
          "@vue/cli-plugin-typescript": "^5.0.8",
          "@vue/cli-service": "^5.0.8",
          typescript: "^4.9.5"
        }
      }, null, 2)
    },
    "/tsconfig.json": {
      code: JSON.stringify({
        compilerOptions: {
          target: "esnext",
          module: "esnext",
          strict: !0,
          jsx: "preserve",
          moduleResolution: "node",
          experimentalDecorators: !0,
          skipLibCheck: !0,
          esModuleInterop: !0,
          allowSyntheticDefaultImports: !0,
          forceConsistentCasingInFileNames: !0,
          useDefineForClassFields: !0,
          sourceMap: !1,
          baseUrl: ".",
          types: ["webpack-env"],
          paths: {
            "@/*": ["src/*"]
          },
          lib: ["esnext", "dom", "dom.iterable", "scripthost"]
        },
        include: ["src/**/*.ts", "src/**/*.tsx", "src/**/*.vue", "tests/**/*.ts", "tests/**/*.tsx"],
        exclude: ["node_modules"]
      }, null, 2)
    }
  },
  main: "/src/App.vue",
  environment: "vue-cli"
}, yi = {
  files: {
    "/src/styles.css": q["/styles.css"],
    "/src/App.vue": {
      code: `<template>
  <div id="app">
    <h1>Hello {{ msg }}</h1>
    <button @click="add">add</button>
    <p>count: {{ count }}</p>
  </div>
</template>

<script>
import Vue from 'vue';

export default Vue.extend({
  name: "App",
  data() {
    return {
      msg: "",
      count: 1,
    };
  },
  methods: {
    add() {
      this.count += 1;
    },
  },
  mounted() {
    this.msg = "world";
  },
});
<\/script>`
    },
    "/src/main.js": {
      code: `import Vue from "vue";
import App from "./App.vue";
import "./styles.css";

Vue.config.productionTip = false;
new Vue({
  render: h => h(App)
}).$mount("#app");
`
    },
    "/public/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0" />
    <link rel="icon" href="<%= BASE_URL %>favicon.ico" />
    <title>codesandbox</title>
  </head>
  <body>
    <noscript>
      <strong
        >We're sorry but codesandbox doesn't work properly without JavaScript
        enabled. Please enable it to continue.</strong
      >
    </noscript>
    <div id="app"></div>
    <!-- built files will be auto injected -->
  </body>
</html>
`
    },
    "/package.json": {
      code: JSON.stringify({
        name: "vue",
        version: "0.1.0",
        private: !0,
        main: "/src/main.js",
        scripts: {
          serve: "vue-cli-service serve",
          build: "vue-cli-service build"
        },
        dependencies: {
          "core-js": "^3.26.1",
          vue: "^2.7.14"
        },
        devDependencies: {
          "@vue/cli-plugin-babel": "^5.0.8",
          "@vue/cli-service": "^5.0.8",
          "vue-template-compiler": "^2.7.14"
        }
      }, null, 2)
    }
  },
  main: "/src/App.vue",
  environment: "vue-cli"
}, gi = {
  files: {
    "/src/styles.css": q["/styles.css"],
    "/src/App.vue": {
      code: `<template>
  <div id="app">
    <h1>Hello {{ msg }}</h1>
    <button @click="add">add</button>
    <p>count: {{ count }}</p>
  </div>
</template>

<script lang="ts">
import Vue from "vue";

export default Vue.extend({
  name: "App",
  data() {
    return {
      msg: "",
      count: 1,
    };
  },
  methods: {
    add() {
      this.count += 1;
    },
  },
  mounted() {
    this.msg = "world";
  },
});
<\/script>`
    },
    "/src/main.ts": {
      code: `import Vue from "vue";
import App from "./App.vue";
import "./styles.css";

Vue.config.productionTip = false;
new Vue({
  render: h => h(App)
}).$mount("#app");
`
    },
    "/public/index.html": {
      code: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0" />
    <link rel="icon" href="<%= BASE_URL %>favicon.ico" />
    <title>codesandbox</title>
  </head>
  <body>
    <noscript>
      <strong
        >We're sorry but codesandbox doesn't work properly without JavaScript
        enabled. Please enable it to continue.</strong
      >
    </noscript>
    <div id="app"></div>
    <!-- built files will be auto injected -->
  </body>
</html>
`
    },
    "/src/shims-vue.d.ts": `declare module '*.vue' {
  import Vue from 'vue'
  export default Vue
}`,
    "/package.json": {
      code: JSON.stringify({
        name: "vue-ts",
        version: "0.1.0",
        private: !0,
        main: "/src/main.ts",
        scripts: {
          serve: "vue-cli-service serve",
          build: "vue-cli-service build"
        },
        dependencies: {
          "core-js": "^3.26.1",
          vue: "^2.7.14"
        },
        devDependencies: {
          "@vue/cli-plugin-babel": "^5.0.8",
          "@vue/cli-plugin-typescript": "^5.0.8",
          "@vue/cli-service": "^5.0.8",
          typescript: "^4.9.3",
          "vue-template-compiler": "^2.7.14"
        }
      }, null, 2)
    },
    "/tsconfig.json": {
      code: JSON.stringify({
        compilerOptions: {
          target: "esnext",
          module: "esnext",
          strict: !0,
          jsx: "preserve",
          moduleResolution: "node",
          skipLibCheck: !0,
          esModuleInterop: !0,
          allowSyntheticDefaultImports: !0,
          forceConsistentCasingInFileNames: !0,
          useDefineForClassFields: !0,
          sourceMap: !0,
          baseUrl: ".",
          types: ["webpack-env"],
          paths: {
            "@/*": ["src/*"]
          },
          lib: ["esnext", "dom", "dom.iterable", "scripthost"]
        },
        include: ["src/**/*.ts", "src/**/*.tsx", "src/**/*.vue", "tests/**/*.ts", "tests/**/*.tsx"],
        exclude: ["node_modules"]
      }, null, 2)
    }
  },
  main: "/src/App.vue",
  environment: "vue-cli"
}, bi = {
  files: O(x({}, q), {
    "/index.html": {
      code: `<!DOCTYPE html>
<html>
<head>
  <title>Parcel Sandbox</title>
  <meta charset="UTF-8" />
  <link rel="stylesheet" href="/styles.css" />
</head>
<body>
  <h1>Hello world</h1>
</body>
</html>`
    },
    "/package.json": {
      code: JSON.stringify({
        dependencies: {},
        main: "/index.html"
      })
    }
  }),
  main: "/index.html",
  environment: "static"
}, ct = {
  react: ui,
  "react-ts": di,
  vue2: yi,
  "vue2-ts": gi,
  vue: zt,
  "vue-ts": Jt,
  vue3: zt,
  "vue3-ts": Jt,
  vanilla: vi,
  "vanilla-ts": hi,
  static: bi,
  angular: ci,
  svelte: pi,
  solid: fi,
  "test-ts": mi,
  node: Xs,
  nextjs: Ks,
  vite: ei,
  "vite-react": ti,
  "vite-react-ts": ni,
  "vite-preact": si,
  "vite-preact-ts": ii,
  "vite-vue": ai,
  "vite-vue-ts": ri,
  "vite-svelte": oi,
  "vite-svelte-ts": li,
  "vite-solid": Qs,
  "vite-lit": Gs,
  astro: Ws
}, Oe = (e) => {
  var a, f, y, m, h, u;
  const n = Ae(e.files), t = Ci({
    template: e.template,
    customSetup: e.customSetup,
    files: n
  });
  let o = Ae((f = (a = e.options) == null ? void 0 : a.visibleFiles) != null ? f : []), l = (y = e.options) != null && y.activeFile ? Yt((m = e.options) == null ? void 0 : m.activeFile, t.files) : void 0;
  o.length === 0 && n && Object.keys(n).forEach((C) => {
    const r = n[C];
    if (typeof r == "string") {
      o.push(C);
      return;
    }
    !l && r.active && (l = C, r.hidden === !0 && o.push(C)), r.hidden || o.push(C);
  }), o.length === 0 && (o = [t.main]), t.entry && !t.files[t.entry] && (t.entry = Yt(t.entry, t.files)), !l && t.main && (l = t.main), (!l || !t.files[l]) && (l = o[0]), o.includes(l) || o.push(l);
  const c = ss(t.files, (h = t.dependencies) != null ? h : {}, (u = t.devDependencies) != null ? u : {}, t.entry);
  return {
    visibleFiles: o.filter((C) => c[C]),
    activeFile: l,
    files: c,
    environment: t.environment,
    shouldUpdatePreview: !0
  };
}, Yt = (e, n) => {
  const t = Ae(n), o = Ae(e);
  if (o in t)
    return o;
  if (!e)
    return null;
  let l = null, c = 0;
  const d = [".js", ".jsx", ".ts", ".tsx"];
  for (; !l && c < d.length; ) {
    const f = `${o.split(".")[0]}${d[c]}`;
    t[f] !== void 0 && (l = f), c++;
  }
  return l;
}, Ci = ({ files: e, template: n, customSetup: t }) => {
  if (!n) {
    if (!t) {
      const l = ct.vanilla;
      return O(x({}, l), {
        files: x(x({}, l.files), Ve(e))
      });
    }
    if (!e || Object.keys(e).length === 0)
      throw new Error("[sandpack-vue3]: without a template, you must pass at least one file");
    return O(x({}, t), {
      files: Ve(e)
    });
  }
  const o = ct[n];
  if (!o)
    throw new Error(`[sandpack-vue3]: invalid template "${n}" provided`);
  return !t && !e ? o : {
    files: Ve(x(x({}, o.files), e)),
    dependencies: x(x({}, o.dependencies), t == null ? void 0 : t.dependencies),
    devDependencies: x(x({}, o.devDependencies), t == null ? void 0 : t.devDependencies),
    entry: Ae((t == null ? void 0 : t.entry) || o.entry),
    main: o.main,
    environment: (t == null ? void 0 : t.environment) || o.environment
  };
}, Ve = (e) => e ? Object.keys(e).reduce((n, t) => (typeof e[t] == "string" ? n[t] = {
  code: e[t]
} : n[t] = e[t], n), {}) : {}, wi = 3e4, yn = Symbol("sandpackStateContext"), gn = A({
  name: "SandpackProvider",
  props: {
    files: {
      type: Object,
      required: !1,
      default: void 0
    },
    template: {
      type: String,
      required: !1,
      default: void 0
    },
    customSetup: {
      type: Object,
      required: !1,
      default: void 0
    },
    theme: {
      type: [String, Object],
      required: !1,
      default: void 0
    },
    options: {
      type: Object,
      required: !1,
      default: void 0
    }
  },
  setup(e, { attrs: n, slots: t }) {
    var oe, de, ve, he, ye, Ne, St;
    const { activeFile: o, visibleFiles: l = [], files: c, environment: d } = Oe(e), a = ((oe = e.options) == null ? void 0 : oe.initMode) || "lazy", f = T(null), y = T(null), m = T({}), h = T(null), u = T(), C = T(), r = pt({
      visibleFiles: l,
      activeFile: o,
      files: c,
      environment: d,
      visibleFilesFromProps: l,
      shouldUpdatePreview: !0,
      editorState: "pristine",
      teamId: e.teamId,
      openFile: D,
      resetFile: J,
      resetAllFiles: W,
      setActiveFile: L,
      updateFile: I,
      updateCurrentFile: K,
      addFile: I,
      closeFile: re,
      deleteFile: F,
      autoReload: (ve = (de = e.options) == null ? void 0 : de.autoReload) != null ? ve : !0,
      reactDevTools: void 0,
      startRoute: (he = e.options) == null ? void 0 : he.startRoute,
      initMode: ((ye = e.options) == null ? void 0 : ye.initMode) || "lazy",
      bundlerState: {
        entry: "",
        transpiledModules: {}
      },
      error: {
        message: ""
      },
      status: (St = (Ne = e.options) == null ? void 0 : Ne.autorun) == null || St ? "initial" : "idle",
      clients: {},
      lazyAnchorRef: null,
      unsubscribeClientListenersRef: {},
      queuedListenersRef: {
        global: {}
      },
      initializeSandpackIframe: ce,
      runSandpack: ie,
      registerBundler: E,
      unregisterBundler: M,
      registerReactDevTools: P,
      addListener: v,
      dispatchMessage: S,
      listen: v,
      dispatch: S
    });
    function P(p) {
      r.reactDevTools = p;
    }
    function D(p) {
      const { visibleFiles: b } = r, N = b.includes(p) ? b : [...b, p];
      r.activeFile = p, r.visibleFiles = N;
    }
    function J(p) {
      const { files: b } = Oe(e);
      r.files = O(x({}, V(r.files)), {
        [p]: b[p]
      });
    }
    function W() {
      const { files: p } = Oe(e);
      r.files = p;
    }
    function L(p) {
      r && r.files[p] && (r.activeFile = p);
    }
    function I(p, b, N = !0) {
      typeof p == "string" && typeof b == "string" ? r.files = O(x({}, V(r.files)), {
        [p]: {
          code: b
        }
      }) : typeof p == "object" && (r.files = Ae(x(x({}, V(r.files)), Ve(p)))), r.shouldUpdatePreview = N;
    }
    function K(p, b = !0) {
      I(r.activeFile, p, b);
    }
    function re(p) {
      if (r.visibleFiles.length === 1)
        return;
      const b = r.visibleFiles.indexOf(p), N = r.visibleFiles.filter((Y) => Y !== p);
      r.activeFile = p === r.activeFile ? b === 0 ? r.visibleFiles[1] : r.visibleFiles[b - 1] : r.activeFile, r.visibleFiles = N;
    }
    function F(p, b = !0) {
      const { visibleFiles: N, files: Y, activeFile: X } = r, ne = x({}, Y);
      delete ne[p];
      const B = N.filter((G) => G !== p);
      if (B.length === 0) {
        const G = Object.keys(Y)[Object.keys(Y).length - 1];
        r.visibleFiles = [G], r.activeFile = G;
      } else
        r.visibleFiles = B, r.activeFile = p === X ? B[B.length - 1] : X;
      r.files = V(ne), r.shouldUpdatePreview = b;
    }
    function S(p, b) {
      var N;
      if (r.status !== "running") {
        console.warn("[sandpack-vue3]: dispatch cannot be called while in idle mode");
        return;
      }
      b ? (N = V(r.clients[b])) == null || N.dispatch(p) : Object.values(r.clients).forEach((Y) => {
        V(Y).dispatch(p);
      });
    }
    function v(p, b) {
      if (b) {
        if (r.clients[b])
          return V(r.clients[b]).listen(p);
        {
          const N = ze();
          return r.queuedListenersRef[b] = r.queuedListenersRef[b] || {}, r.unsubscribeClientListenersRef[b] = r.unsubscribeClientListenersRef[b] || {}, r.queuedListenersRef[b][N] = p, () => {
            r.queuedListenersRef[b][N] ? delete r.queuedListenersRef[b][N] : r.unsubscribeClientListenersRef[b][N] && (r.unsubscribeClientListenersRef[b][N](), delete r.unsubscribeClientListenersRef[b][N]);
          };
        }
      } else {
        const N = ze();
        r.queuedListenersRef.global[N] = p;
        const X = Object.values(r.clients).map((B) => V(B).listen(p));
        return () => {
          X.forEach((B) => B()), delete r.queuedListenersRef.global[N], Object.values(r.unsubscribeClientListenersRef).forEach((B) => {
            var Z;
            (Z = B == null ? void 0 : B[N]) == null || Z.call(B);
          });
        };
      }
    }
    function E(p, b, N) {
      return ke(this, null, function* () {
        m.value[b] = {
          iframe: p,
          clientPropsOverride: N
        }, r.status === "running" && (yield $(p, b, N));
      });
    }
    function M(p) {
      var X, ne;
      const b = V(r.clients[p]);
      b && (b.destroy(), (X = b.iframe.contentWindow) == null || X.location.replace("about:blank"), b.iframe.removeAttribute("src"), delete r.clients[p]), delete m.value[p], h.value && clearTimeout(h.value), Object.values((ne = r.unsubscribeClientListenersRef[p]) != null ? ne : {}).forEach((B) => {
        Object.values(B).forEach((G) => G());
      });
      const Y = Object.keys(r.clients).length > 0 ? "running" : "idle";
      r.status = Y;
    }
    function j(p) {
      p.type === "start" ? r.error = null : p.type === "state" ? r.bundlerState = p.state : p.type === "done" && !p.compilatonError || p.type === "connected" ? (h.value && clearTimeout(h.value), r.error = null) : p.type === "action" && p.action === "show-error" ? r.error = os(p) : p.type === "action" && p.action === "notification" && p.notificationType === "error" && (r.error = {
        message: p.title
      });
    }
    function $(p, b, N) {
      return ke(this, null, function* () {
        var G, De, xt, Et, kt, Tt, At, Lt, Nt, Ot, Mt;
        r.clients[b] && r.clients[b].destroy();
        const Y = (G = e == null ? void 0 : e.customSetup) != null ? G : {
          npmRegistries: []
        }, X = (xt = (De = e == null ? void 0 : e.options) == null ? void 0 : De.bundlerTimeOut) != null ? xt : wi;
        h.value && clearTimeout(h.value), typeof u.value != "function" && (h.value = setTimeout(() => {
          ee(), r.status = "timeout";
        }, X));
        const B = yield is(p, {
          files: V(r.files),
          template: V(r.environment)
        }, {
          externalResources: (Et = e.options) == null ? void 0 : Et.externalResources,
          bundlerURL: (kt = e.options) == null ? void 0 : kt.bundlerURL,
          startRoute: (At = N == null ? void 0 : N.startRoute) != null ? At : (Tt = e.options) == null ? void 0 : Tt.startRoute,
          fileResolver: (Lt = e.options) == null ? void 0 : Lt.fileResolver,
          skipEval: (Ot = (Nt = e.options) == null ? void 0 : Nt.skipEval) != null ? Ot : !1,
          logLevel: (Mt = e.options) == null ? void 0 : Mt.logLevel,
          showOpenInCodeSandbox: !1,
          showErrorScreen: !0,
          showLoadingScreen: !0,
          reactDevTools: r.reactDevTools,
          customNpmRegistries: Y == null ? void 0 : Y.npmRegistries,
          teamId: e.teamId,
          sandboxId: e.sandboxId
        });
        typeof u.value != "function" && (u.value = B.listen(j)), r.unsubscribeClientListenersRef[b] = r.unsubscribeClientListenersRef[b] || {}, r.queuedListenersRef[b] && (Object.keys(r.queuedListenersRef[b]).forEach((Be) => {
          const Xe = r.queuedListenersRef[b][Be], Qe = B.listen(Xe);
          r.unsubscribeClientListenersRef[b][Be] = Qe;
        }), r.queuedListenersRef[b] = {}), Object.entries(r.queuedListenersRef.global).forEach(([Be, Xe]) => {
          const Qe = B.listen(Xe);
          r.unsubscribeClientListenersRef[b][Be] = Qe;
        }), r.clients[b] = B;
      });
    }
    function ee() {
      Object.keys(r.clients).map(M), typeof u.value == "function" && (u.value(), u.value = void 0);
    }
    function ie() {
      return ke(this, null, function* () {
        yield Promise.all(Object.entries(m.value).map((Y) => ke(this, [Y], function* ([p, { iframe: b, clientPropsOverride: N = {} }]) {
          yield $(b, p, N);
        }))), r.error = null, r.status = "running";
      });
    }
    function ce() {
      var N, Y, X, ne, B;
      if (!((Y = (N = e.options) == null ? void 0 : N.autorun) != null ? Y : !0))
        return;
      const b = (ne = (X = e.options) == null ? void 0 : X.initModeObserverOptions) != null ? ne : {
        rootMargin: "1000px 0px"
      };
      f.value && r.lazyAnchorRef && ((B = f.value) == null || B.unobserve(r.lazyAnchorRef)), r.lazyAnchorRef && r.initMode === "lazy" ? (f.value = new IntersectionObserver((Z) => {
        var G;
        Z.some((De) => De.isIntersecting) && (y.value = setTimeout(() => {
          ie();
        }, 50), r.lazyAnchorRef && ((G = f.value) == null || G.unobserve(r.lazyAnchorRef)));
      }, b), f.value.observe(r.lazyAnchorRef)) : r.lazyAnchorRef && r.initMode === "user-visible" ? (f.value = new IntersectionObserver((Z) => {
        Z.some((G) => G.isIntersecting) ? y.value = setTimeout(() => {
          ie();
        }, 50) : (y.value && clearTimeout(y.value), Object.keys(r.clients).map(M), ee());
      }, b), f.value.observe(r.lazyAnchorRef)) : y.value = setTimeout(() => ie(), 50);
    }
    return Je(yn, r), H([() => r.files, () => r.environment, () => r.shouldUpdatePreview, () => r.reactDevTools, () => r.status, () => {
      var p;
      return (p = e.options) == null ? void 0 : p.recompileMode;
    }, () => {
      var p;
      return (p = e.options) == null ? void 0 : p.recompileDelay;
    }], () => {
      var Y, X, ne, B;
      const { environment: p } = Oe(e), b = (X = (Y = e.options) == null ? void 0 : Y.recompileMode) != null ? X : "delayed", N = (B = (ne = e.options) == null ? void 0 : ne.recompileDelay) != null ? B : 500;
      if (!(r.status !== "running" || !r.shouldUpdatePreview) && (p !== r.environment && (r.environment = p, Object.entries(r.clients).forEach(([Z, G]) => {
        E(V(G).iframe, Z);
      })), b === "immediate" && Object.values(r.clients).forEach((Z) => {
        Z.status === "done" && V(Z).updateSandbox({
          files: x({}, V(r.files)),
          template: V(r.environment)
        });
      }), b === "delayed")) {
        if (typeof window > "u")
          return;
        window.clearTimeout(C.value), C.value = window.setTimeout(() => {
          Object.values(r.clients).forEach((Z) => {
            Z.status === "done" && V(Z).updateSandbox({
              files: x({}, V(r.files)),
              template: V(r.environment)
            });
          });
        }, N);
      }
    }, {
      immediate: !0,
      deep: !0
    }), H([() => {
      var p;
      return (p = e.options) == null ? void 0 : p.initMode;
    }, () => r.initMode, () => {
      var p;
      return (p = e.options) == null ? void 0 : p.autorun;
    }, () => {
      var p;
      return (p = e.options) == null ? void 0 : p.initModeObserverOptions;
    }, () => r.environment, () => r.files, () => r.reactDevTools], () => {
      a !== r.initMode && (r.initMode = a, ce());
    }, {
      immediate: !0,
      deep: !0
    }), H([() => e.options, () => e.template, () => e == null ? void 0 : e.files, () => e.customSetup], () => {
      const p = Oe(e);
      r.activeFile = p.activeFile, r.visibleFiles = p.visibleFiles, r.visibleFilesFromProps = p.visibleFiles, r.files = p.files, r.environment = p.environment;
      const b = ns(p.files, V(r.files)) ? "pristine" : "dirty";
      b !== r.editorState && (r.editorState = b);
    }, {
      deep: !0
    }), xe(() => {
      typeof u.value == "function" && u.value(), h.value && clearTimeout(h.value), C.value && clearTimeout(C.value), y.value && clearTimeout(y.value), f.value && f.value.disconnect(), Ie(() => {
        ce();
      });
    }), () => {
      var p;
      return s(Ys, {
        classes: (p = e.options) == null ? void 0 : p.classes
      }, {
        default: () => [s(ki, {
          style: (n == null ? void 0 : n.style) || {},
          class: (n == null ? void 0 : n.class) || "",
          theme: e.theme
        }, {
          default: () => [t.default ? t.default() : null]
        })]
      });
    };
  }
});
function z() {
  let e = mt(yn, void 0);
  return e || (e = {
    files: {},
    dispatch: () => {
    },
    listen: () => {
    }
  }, console.error('[sandpack-vue3]: "useSandpack" must be wrapped by a "SandpackProvider"')), {
    sandpack: e,
    dispatch: e.dispatch,
    listen: e.listen
  };
}
const bn = () => {
  const { sandpack: e } = z(), n = k(() => {
    var o;
    return (o = e == null ? void 0 : e.files[e.activeFile]) == null ? void 0 : o.code;
  }), t = k(() => {
    var o, l;
    return (l = (o = e == null ? void 0 : e.files[e.activeFile]) == null ? void 0 : o.readOnly) != null ? l : !1;
  });
  return {
    code: n,
    readOnly: t,
    updateCode: e.updateCurrentFile
  };
}, Cn = 200, Si = (e) => {
  let n, t;
  const { sandpack: o, listen: l } = z(), c = T("LOADING"), d = () => {
    n && n(), n = l((a) => {
      a.type === "start" && a.firstLoad === !0 && (c.value = "LOADING"), a.type === "done" && (c.value = c.value === "LOADING" ? "PRE_FADING" : "HIDDEN");
    }, e.clientId);
  };
  return Qn(() => {
    c.value = "LOADING";
  }), ae(() => {
    n && n();
  }), se(() => {
    n && n();
  }), H(() => o.status, () => {
    o.status !== "running" && (c.value = "HIDDEN"), o.status === "timeout" && (c.value = "TIMEOUT"), o.status === "idle" && d();
  }, {
    immediate: !0
  }), H(() => e.clientId, d, {
    immediate: !0
  }), H([c, () => e.loading], () => {
    t && clearTimeout(t), c.value === "PRE_FADING" && !e.loading ? c.value = "FADING" : c.value === "FADING" && (t && clearTimeout(t), t = setTimeout(() => {
      c.value = "HIDDEN";
    }, Cn));
  }, {
    deep: !0,
    immediate: !0
  }), c;
}, xi = (e) => {
  const { dispatch: n } = z();
  return {
    refresh: () => n({
      type: "refresh"
    }, e),
    back: () => n({
      type: "urlback"
    }, e),
    forward: () => n({
      type: "urlforward"
    }, e)
  };
}, Ei = g, wn = Symbol("sandpackThemeProvider"), ki = A({
  name: "SandpackThemeProvider",
  props: {
    theme: {
      type: [String, Object],
      required: !1,
      default: void 0
    },
    mode: {
      type: String,
      required: !1,
      default: "light"
    }
  },
  setup(e, { slots: n, attrs: t }) {
    const { theme: o, id: l = "", mode: c = "light" } = Vt(e.theme), d = R(), a = T({}), f = pt({
      theme: o,
      id: l,
      mode: c
    });
    return f.theme && (a.value = _t(l, Ht(o)), Je(wn, f)), H(() => e.theme, () => {
      const { theme: m, id: h } = Vt(e.theme);
      e.theme && (f.id = h, f.theme = m, a.value = _t(h, Ht(m)));
    }, {
      immediate: !0
    }), () => s("div", {
      class: d("wrapper", [a.value.toString(), Ei(), (t == null ? void 0 : t.class) || ""])
    }, [n.default ? n.default() : null]);
  }
}), Hl = (e, n = Symbol()) => {
  if (!e || typeof e != "object")
    throw new Error("context can only be object: Record<string, any>");
  const t = pt(x({}, e)), o = A({
    name: "ContextProvider",
    inheritAttrs: !1,
    setup(l, { slots: c }) {
      return Je(n, t), () => {
        var d;
        return (d = c.default) == null ? void 0 : d.call(c);
      };
    }
  });
  return {
    state: t,
    Provider: o
  };
}, mt = (e = Symbol(), n) => {
  const t = es(e, n);
  return t == null ? n : t;
}, Sn = () => mt(wn, {
  theme: Te,
  id: "light",
  mode: "light"
}), vt = (e) => {
  const { sandpack: n, listen: t, dispatch: o } = z(), l = T(null), c = T(ze());
  return xe(() => {
    const a = l.value;
    a !== null && n.registerBundler(a, c.value, e);
  }), ae(() => {
    n && c.value && n.unregisterBundler(c.value);
  }), se(() => {
    n && c.value && n.unregisterBundler(c.value);
  }), {
    sandpack: n,
    getClient: () => {
      const { clients: a } = n;
      return V(a[c.value] || null);
    },
    clientId: c,
    iframe: l,
    listen: (a) => t(a, c.value),
    dispatch: (a) => o(a, c.value)
  };
}, Ye = (e) => {
  const { dispatch: n } = z();
  return {
    restart: () => n({
      type: "shell/restart"
    }, e),
    openPreview: () => n({
      type: "shell/openPreview"
    }, e)
  };
}, Ti = ({ langSupport: e, highlightTheme: n, code: t = "" }) => {
  const o = e.language.parser.parse(t);
  let l = 0;
  const c = [], d = (a, f) => {
    if (a > l) {
      const y = t.slice(l, a);
      c.push(f ? s("span", {
        key: `${a}${l}`,
        class: f
      }, [y]) : y), l = a;
    }
  };
  return bs(o, n, (a, f, y) => {
    d(a, ""), d(f, y);
  }), l < t.length && (t != null && t.includes(`
`)) && c.push(`

`), c;
};
function Ai(e) {
  let n;
  const t = Sn(), o = _s(e.filePath, e.fileType, e.additionalLanguages), l = Hs(o, e.additionalLanguages), c = $s(t.theme), d = Ti({
    langSupport: l,
    highlightTheme: c,
    code: e.code
  }), a = T(e.code), f = T(), y = T(), m = T(!1), h = k(() => e.initMode || "lazy"), u = k(() => {
    const L = e.decorators ? [...e.decorators] : [];
    return L && L.sort((I, K) => I.line - K.line);
  }), C = T(h.value === "immediate"), r = T([]), P = T([]);
  Fs(f, ([{ isIntersecting: L }]) => {
    m.value = L;
  }, {
    rootMargin: "600px 0px",
    threshold: 0.2
  });
  const D = () => {
    !f.value || !C.value || (W(), n = setTimeout(J));
  }, J = () => {
    const L = [{
      key: "Tab",
      run: (S) => {
        var E, M;
        ks(S);
        const v = e.extensionsKeymap.find(({ key: j }) => j === "Tab");
        return (M = (E = v == null ? void 0 : v.run) == null ? void 0 : E.call(v, S)) != null ? M : !0;
      }
    }, {
      key: "Shift-Tab",
      run: ({ state: S, dispatch: v }) => {
        var M, j;
        Ts({
          state: S,
          dispatch: v
        });
        const E = e.extensionsKeymap.find(({ key: $ }) => $ === "Shift-Tab");
        return (j = (M = E == null ? void 0 : E.run) == null ? void 0 : M.call(E, F)) != null ? j : !0;
      }
    }, {
      key: "Escape",
      run: () => (e.readOnly || f.value && f.value.focus(), !0)
    }, {
      key: "mod-Backspace",
      run: As
    }], I = [ms(), Ss(), cs(), ...e.extensions, It.of([...us, ...xs, ...Es, ...L, ...e.extensionsKeymap]), l, qs(), fs(c)];
    e.readOnly ? (I.push(Cs.readOnly.of(!0)), I.push(He.editable.of(!1))) : (I.push(ps()), I.push(vs())), u.value && I.push(Vs(u.value)), e.wrapContent && I.push(He.lineWrapping), e.showLineNumbers && I.push(hs()), e.showInlineErrors && I.push(Us());
    const K = f.value, re = K.querySelector(".sp-pre-placeholder");
    re && K.removeChild(re);
    const F = new He({
      doc: e.code,
      extensions: I,
      parent: K,
      dispatch: (S) => {
        var v;
        if (F.update([S]), S.docChanged) {
          const E = S.newDoc.sliceString(0, S.newDoc.length);
          a.value = E, (v = e.onCodeUpdate) == null || v.call(e, E);
        }
      }
    });
    F.contentDOM.setAttribute("data-gramm", "false"), F.contentDOM.setAttribute("data-lt-active", "false"), F.contentDOM.setAttribute("aria-label", e.filePath ? `Code Editor for ${Ue(e.filePath)}` : "Code Editor"), e.readOnly ? F.contentDOM.classList.add("cm-readonly") : F.contentDOM.setAttribute("tabIndex", "-1"), y.value = F;
  }, W = () => {
    var L;
    (L = y.value) == null || L.destroy(), n && clearTimeout(n);
  };
  return H([() => e.readOnly, () => e.showLineNumbers, () => e.showInlineErrors, () => e.wrapContent, () => e.readOnly, () => t.id, C, u], D, {
    immediate: !0
  }), H([() => e.initMode, m], () => {
    (h.value === "lazy" || h.value === "user-visible") && m.value && (C.value = !0);
  }), H([() => e.extensions, () => e.extensionsKeymap], () => {
    const L = y.value, I = !Ut(e.extensions, r.value) || !Ut(e.extensionsKeymap, P.value);
    L && I && (L.dispatch({
      effects: Bt.appendConfig.of(e.extensions || [])
    }), L.dispatch({
      effects: Bt.appendConfig.of(It.of([...e.extensionsKeymap || []]))
    }), r.value = e.extensions, P.value = e.extensionsKeymap);
  }, {
    immediate: !0
  }), xe(() => {
    Ie(D);
  }), se(W), {
    internalCode: a,
    shouldInitEditor: C,
    wrapperRef: f,
    cmView: y,
    languageExtension: o,
    syntaxHighlightRender: d
  };
}
const Le = g, le = g, Ce = g, Li = g;
vn({
  "0%": {
    opacity: 0,
    transform: "translateY(4px)"
  },
  "100%": {
    opacity: 1,
    transform: "translateY(0)"
  }
});
const Fe = g, xn = g, ut = g, ge = g, Zt = g, Wt = g, En = g, Gt = g, Ni = g, Ze = A({
  name: "CodeMirror",
  props: Ls,
  setup(e, n) {
    let t;
    const { cmView: o, wrapperRef: l, languageExtension: c, internalCode: d, syntaxHighlightRender: a } = Ai(e), f = R(), { listen: y } = z(), m = (u) => {
      u.key === "Enter" && o.value && (u.preventDefault(), o.value.contentDOM.focus());
    };
    n.expose({
      getCodemirror: () => o.value
    }), xe(() => {
      Ie(() => {
        o.value && e.editorState === "dirty" && window.matchMedia("(min-width: 768px)").matches && o.value.contentDOM.focus();
      });
    }), H(() => e.code, (u) => {
      if (o.value && typeof u == "string" && u !== d.value) {
        const C = o.value, r = C.state.selection.ranges.some(({ to: D, from: J }) => D > u.length || J > u.length) ? ws.cursor(u.length) : C.state.selection, P = {
          from: 0,
          to: C.state.doc.length,
          insert: u
        };
        C.dispatch({
          changes: P,
          selection: r
        });
      }
    }, {
      immediate: !0
    }), H([() => e.showInlineErrors], () => {
      e.showInlineErrors && (t && t(), t = y((u) => {
        const C = o.value;
        u.type === "success" ? C == null || C.dispatch({
          annotations: [new Pt("remove-errors", !0)]
        }) : u.type === "action" && u.action === "show-error" && u.path === e.filePath && u.line && (C == null || C.dispatch({
          annotations: [new Pt("show-error", u.line)]
        })), u.type === "action" && u.action === "show-error" && u.line && (C == null || C.dispatch({
          annotations: [{
            type: "error",
            value: u.line
          }]
        }));
      }));
    }, {
      immediate: !0
    });
    const h = () => {
      let u = 4;
      return e.showLineNumbers && (u += 6), e.readOnly || (u += 1), `var(--${U}-space-${u})`;
    };
    return ae(() => {
      t && t();
    }), se(() => {
      t && t();
    }), () => e.readOnly ? s(Q, null, [s("pre", {
      ref: l,
      class: f("cm", [f(e.editorState), f(c), Gt, Wt]),
      translate: "no"
    }, [s("code", {
      class: f("pre-placeholder", [Zt]),
      style: {
        marginLeft: h()
      }
    }, [a])]), e.readOnly && e.showReadOnly && s("span", {
      class: f("read-only", [Ni])
    }, [w("Read-only")])]) : s("div", {
      ref: l,
      "aria-autocomplete": "list",
      "aria-label": e.filePath ? `Code Editor for ${Ue(e.filePath)}` : "Code Editor",
      "aria-multiline": "true",
      onKeydown: m,
      class: f("cm", [f(e.editorState), f(c), Gt, Wt]),
      role: "textbox",
      tabindex: 0,
      translate: "no"
    }, [s("pre", {
      class: f("pre-placeholder", [Zt]),
      style: {
        marginLeft: h()
      }
    }, [a])]);
  }
}), te = A({
  props: {
    fill: {
      type: String,
      required: !1,
      default: "currentColor"
    },
    height: {
      type: String,
      required: !1,
      default: "16"
    },
    viewBox: {
      type: String,
      required: !1,
      default: "0 0 16 16"
    },
    stroke: {
      type: String,
      required: !1,
      default: void 0
    },
    width: {
      type: String,
      required: !1,
      default: "16"
    },
    xmlns: {
      type: String,
      required: !1,
      default: "http://www.w3.org/2000/svg"
    }
  },
  setup(e, { slots: n }) {
    return () => s("svg", e, [n.default ? n.default() : null]);
  }
}), We = () => s(te, {
  fill: "none",
  stroke: "currentColor"
}, {
  default: () => [s("title", null, [w("Restart script")]), s("path", {
    d: "M8 2C4.68629 2 2 4.68629 2 8C2 10.0946 3.07333 11.9385 4.7 13.0118",
    "stroke-linecap": "round"
  }, null), s("path", {
    d: "M14.0005 7.9998C14.0005 5.82095 12.8391 3.91335 11.1016 2.8623",
    "stroke-linecap": "round"
  }, null), s("path", {
    d: "M14.0003 2.3335H11.167C10.8908 2.3335 10.667 2.55735 10.667 2.8335V5.66683",
    "stroke-linecap": "round"
  }, null), s("path", {
    d: "M1.99967 13.6665L4.83301 13.6665C5.10915 13.6665 5.33301 13.4426 5.33301 13.1665L5.33301 10.3332",
    "stroke-linecap": "round"
  }, null), s("path", {
    d: "M10 10L12 12L10 14",
    "stroke-linecap": "round",
    "stroke-linejoin": "round"
  }, null), s("path", {
    d: "M14.667 14L12.667 14",
    "stroke-linecap": "round",
    "stroke-linejoin": "round"
  }, null)]
}), kn = () => s(te, null, {
  default: () => [s("title", null, [w("Run sandbox")]), s("path", {
    d: "M11.0792 8.1078C11.2793 8.25007 11.27 8.55012 11.0616 8.67981L6.02535 11.8135C5.79638 11.956 5.5 11.7913 5.5 11.5216L5.5 8.40703L5.5 4.80661C5.5 4.52735 5.81537 4.36463 6.04296 4.52647L11.0792 8.1078Z"
  }, null)]
}), Oi = () => s(te, null, {
  default: () => [s("title", null, [w("Click to go back")]), s("path", {
    d: "M9.64645 12.3536C9.84171 12.5488 10.1583 12.5488 10.3536 12.3536C10.5488 12.1583 10.5488 11.8417 10.3536 11.6464L9.64645 12.3536ZM10.3536 4.35355C10.5488 4.15829 10.5488 3.84171 10.3536 3.64644C10.1583 3.45118 9.84171 3.45118 9.64645 3.64644L10.3536 4.35355ZM6.07072 7.92929L5.71716 7.57573L6.07072 7.92929ZM10.3536 11.6464L6.42427 7.71716L5.71716 8.42426L9.64645 12.3536L10.3536 11.6464ZM6.42427 8.28284L10.3536 4.35355L9.64645 3.64644L5.71716 7.57573L6.42427 8.28284ZM6.42427 7.71716C6.58048 7.87337 6.58048 8.12663 6.42427 8.28284L5.71716 7.57573C5.48285 7.81005 5.48285 8.18995 5.71716 8.42426L6.42427 7.71716Z"
  }, null)]
}), Mi = () => s(te, null, {
  default: () => [s("title", null, [w("Click to go forward")]), s("path", {
    d: "M6.35355 3.64645C6.15829 3.45118 5.84171 3.45118 5.64645 3.64645C5.45118 3.84171 5.45118 4.15829 5.64645 4.35355L6.35355 3.64645ZM5.64645 11.6464C5.45118 11.8417 5.45118 12.1583 5.64645 12.3536C5.84171 12.5488 6.15829 12.5488 6.35355 12.3536L5.64645 11.6464ZM9.92929 8.07071L10.2828 8.42426L9.92929 8.07071ZM5.64645 4.35355L9.57574 8.28284L10.2828 7.57574L6.35355 3.64645L5.64645 4.35355ZM9.57574 7.71716L5.64645 11.6464L6.35355 12.3536L10.2828 8.42426L9.57574 7.71716ZM9.57574 8.28284C9.41952 8.12663 9.41953 7.87337 9.57574 7.71716L10.2828 8.42426C10.5172 8.18995 10.5172 7.81005 10.2828 7.57574L9.57574 8.28284Z"
  }, null)]
}), Tn = () => s(te, null, {
  default: () => [s("title", null, [w("Refresh preview")]), s("path", {
    "clip-rule": "evenodd",
    d: "M3.83325 7.99992C3.83325 5.69867 5.69853 3.83325 7.99934 3.83325C9.81246 3.83325 11.3563 4.99195 11.9285 6.61097C11.9396 6.6425 11.9536 6.67221 11.97 6.69992H8.80005C8.52391 6.69992 8.30005 6.92378 8.30005 7.19992C8.30005 7.47606 8.52391 7.69992 8.80005 7.69992H12.5667C12.8981 7.69992 13.1667 7.43129 13.1667 7.09992V3.33325C13.1667 3.05711 12.9429 2.83325 12.6667 2.83325C12.3906 2.83325 12.1667 3.05711 12.1667 3.33325V4.94608C11.2268 3.66522 9.7106 2.83325 7.99934 2.83325C5.14613 2.83325 2.83325 5.14651 2.83325 7.99992C2.83325 10.8533 5.14613 13.1666 7.99934 13.1666C9.91218 13.1666 11.5815 12.1266 12.474 10.5836C12.6123 10.3446 12.5306 10.0387 12.2915 9.90044C12.0525 9.76218 11.7466 9.84387 11.6084 10.0829C10.8873 11.3296 9.54072 12.1666 7.99934 12.1666C5.69853 12.1666 3.83325 10.3012 3.83325 7.99992Z",
    "fill-rule": "evenodd"
  }, null)]
}), Fi = () => s(te, {
  fill: "none",
  stroke: "currentColor"
}, {
  default: () => [s("title", null, [w("Clean")]), s("circle", {
    cx: "7.99998",
    cy: "8.00004",
    r: "4.66667",
    "stroke-linecap": "round"
  }, null), s("path", {
    d: "M4.66669 4.66663L11.3334 11.3333"
  }, null)]
}), ji = () => s(te, {
  fill: "none",
  stroke: "currentColor"
}, {
  default: () => [s("title", null, [w("Open on CodeSandbox")]), s("path", {
    d: "M6.66665 3.33337H4.33331C3.78103 3.33337 3.33331 3.78109 3.33331 4.33337V11.6667C3.33331 12.219 3.78103 12.6667 4.33331 12.6667H11.6666C12.2189 12.6667 12.6666 12.219 12.6666 11.6667V9.33337",
    "stroke-linecap": "round"
  }, null), s("path", {
    d: "M10 3.33337H12.5667C12.6219 3.33337 12.6667 3.37815 12.6667 3.43337V6.00004",
    "stroke-linecap": "round"
  }, null), s("path", {
    d: "M7.33331 8.66668L12.5333 3.46667",
    "stroke-linecap": "round"
  }, null)]
}), Ri = () => s(te, null, {
  default: () => [s("title", null, [w("Directory")]), s("path", {
    d: "M12.5526 12.6667H3.66675C3.2922 12.6667 2.96575 12.4608 2.79442 12.156L3.81072 8.0908C3.92201 7.64563 4.32199 7.33333 4.78086 7.33333H13.386C14.0365 7.33333 14.5139 7.94472 14.3561 8.57587L13.5228 11.9092C13.4115 12.3544 13.0115 12.6667 12.5526 12.6667Z",
    fill: "currentColor"
  }, null), s("path", {
    d: "M13.3334 6.63333V6.33333C13.3334 5.78105 12.8857 5.33333 12.3334 5.33333H8.30286C8.10543 5.33333 7.91242 5.2749 7.74816 5.16538L6.25201 4.16795C6.08774 4.05844 5.89473 4 5.69731 4H3.66675C3.11446 4 2.66675 4.44772 2.66675 5L2.66675 11.6667C2.66675 12.219 3.11446 12.6667 3.66675 12.6667H12.5526C13.0115 12.6667 13.4115 12.3544 13.5228 11.9092L14.3561 8.57587C14.5139 7.94472 14.0365 7.33333 13.386 7.33333H4.78086C4.32199 7.33333 3.92201 7.64563 3.81072 8.0908L2.75008 12.3333",
    fill: "none",
    stroke: "currentColor",
    "stroke-linecap": "round"
  }, null)]
}), Ii = () => s(te, null, {
  default: () => [s("title", null, [w("Directory")]), s("path", {
    d: "M12.3334 12.6667H3.66675C3.11446 12.6667 2.66675 12.219 2.66675 11.6667V5C2.66675 4.44772 3.11446 4 3.66675 4H5.69731C5.89473 4 6.08774 4.05844 6.25201 4.16795L7.74816 5.16538C7.91242 5.2749 8.10543 5.33333 8.30286 5.33333H12.3334C12.8857 5.33333 13.3334 5.78105 13.3334 6.33333V11.6667C13.3334 12.219 12.8857 12.6667 12.3334 12.6667Z",
    fill: "currentColor",
    stroke: "currentColor",
    "stroke-linecap": "round"
  }, null)]
}), Di = () => s(te, {
  fill: "currentColor"
}, {
  default: () => [s("title", null, [w("File")]), s("path", {
    "clip-rule": "evenodd",
    d: "M4.5 4.33325C4.5 4.05711 4.72386 3.83325 5 3.83325H8.16675V6.56659C8.16675 6.89795 8.43538 7.16658 8.76675 7.16658H11.5V12.3333C11.5 12.6094 11.2761 12.8333 11 12.8333H5C4.72386 12.8333 4.5 12.6094 4.5 12.3333V4.33325ZM12.5 6.67568C12.5001 6.67265 12.5001 6.66962 12.5001 6.66658C12.5001 6.66355 12.5001 6.66052 12.5 6.65749V6.41413C12.5 6.01631 12.342 5.63478 12.0607 5.35347L9.97978 3.27259C9.69848 2.99129 9.31694 2.83325 8.91912 2.83325H8.66675H5C4.17157 2.83325 3.5 3.50483 3.5 4.33325V12.3333C3.5 13.1617 4.17157 13.8333 5 13.8333H11C11.8284 13.8333 12.5 13.1617 12.5 12.3333V6.67568ZM9.16675 3.89888C9.20518 3.92078 9.24085 3.94787 9.27267 3.9797L11.3536 6.06058C11.3854 6.09243 11.4125 6.12813 11.4344 6.16658H9.16675V3.89888Z",
    "fill-rule": "evenodd"
  }, null)]
}), Bi = () => s(te, {
  stroke: "currentColor"
}, {
  default: () => [s("title", null, [w("Close file")]), s("path", {
    d: "M12 4L4 12",
    "stroke-linecap": "round"
  }, null), s("path", {
    d: "M4 4L12 12",
    "stroke-linecap": "round"
  }, null)]
}), ht = () => s(te, null, {
  default: () => [s("title", null, [w("Open browser console")]), s("path", {
    d: "M5.65871 3.62037C5.44905 3.44066 5.1334 3.46494 4.95368 3.6746C4.77397 3.88427 4.79825 4.19992 5.00792 4.37963L5.65871 3.62037ZM5.00792 11.6204C4.79825 11.8001 4.77397 12.1157 4.95368 12.3254C5.1334 12.5351 5.44905 12.5593 5.65871 12.3796L5.00792 11.6204ZM9.9114 7.92407L10.2368 7.54445L9.9114 7.92407ZM5.00792 4.37963L9.586 8.3037L10.2368 7.54445L5.65871 3.62037L5.00792 4.37963ZM9.586 7.6963L5.00792 11.6204L5.65871 12.3796L10.2368 8.45555L9.586 7.6963ZM9.586 8.3037C9.39976 8.14407 9.39976 7.85594 9.586 7.6963L10.2368 8.45555C10.5162 8.2161 10.5162 7.7839 10.2368 7.54445L9.586 8.3037Z"
  }, null), s("path", {
    d: "M10 11.5C9.72386 11.5 9.5 11.7239 9.5 12C9.5 12.2761 9.72386 12.5 10 12.5V11.5ZM14.6667 12.5C14.9428 12.5 15.1667 12.2761 15.1667 12C15.1667 11.7239 14.9428 11.5 14.6667 11.5V12.5ZM10 12.5H14.6667V11.5H10V12.5Z"
  }, null)]
}), Pi = () => s(te, {
  viewBox: "0 0 48 48"
}, {
  default: () => [s("title", null, [w("Sign in")]), s("path", {
    d: "M9 42q-1.2 0-2.1-.9Q6 40.2 6 39V9q0-1.2.9-2.1Q7.8 6 9 6h14.55v3H9v30h14.55v3Zm24.3-9.25-2.15-2.15 5.1-5.1h-17.5v-3h17.4l-5.1-5.1 2.15-2.15 8.8 8.8Z"
  }, null)]
}), qi = () => s(te, {
  viewBox: "0 0 48 48"
}, {
  default: () => [s("title", null, [w("Sign out")]), s("path", {
    d: "M9 42q-1.2 0-2.1-.9Q6 40.2 6 39V9q0-1.2.9-2.1Q7.8 6 9 6h14.55v3H9v30h14.55v3Zm24.3-9.25-2.15-2.15 5.1-5.1h-17.5v-3h17.4l-5.1-5.1 2.15-2.15 8.8 8.8Z"
  }, null)]
}), $i = g, _i = g, Hi = g, Vi = g, An = A({
  name: "FileTabs",
  props: {
    closableTabs: {
      type: Boolean,
      required: !1,
      default: void 0
    }
  },
  setup(e, { attrs: n }) {
    const { sandpack: t } = z(), o = R(), l = (d) => {
      d.stopPropagation();
      const a = d.target.closest("[data-active]"), f = a == null ? void 0 : a.getAttribute("title");
      f && t.closeFile(f);
    }, c = (d) => {
      const a = Ue(d), f = (t.visibleFiles || []).reduce((y, m) => (m === d || Ue(m) === a && y.push(m), y), []);
      return f.length === 0 ? a : js(d, f);
    };
    return () => s("div", {
      class: o("tabs", [$i, (n == null ? void 0 : n.class) || ""]),
      translate: "no"
    }, [s("div", {
      "aria-label": "Select active file",
      class: o("tabs-scrollable-container", [_i]),
      role: "tablist"
    }, [(t.visibleFiles || []).map((d) => s("button", {
      key: d,
      "aria-selected": d === t.activeFile,
      class: o("tab-button", [le, Vi]),
      "data-active": d === t.activeFile,
      onClick: () => t.setActiveFile(d),
      role: "tab",
      title: d,
      type: "button",
      style: {
        display: "flex !important"
      }
    }, [c(d), e.closableTabs && t.visibleFiles.length > 1 && s("span", {
      class: o("close-button", [Hi]),
      onClick: l
    }, [s(Bi, null, null)])]))])]);
  }
}), Se = A({
  name: "RoundedButton",
  props: {
    title: {
      type: String,
      required: !1,
      default: ""
    },
    onClick: {
      type: Function,
      required: !1,
      default: () => {
      }
    }
  },
  setup(e, { slots: n, attrs: t }) {
    const o = R();
    return () => s("button", {
      title: e.title,
      class: o("button", [o("icon-standalone"), le, Le, Ce, (t == null ? void 0 : t.class) || ""]),
      onClick: e.onClick,
      type: "button"
    }, [n.default ? n.default() : null]);
  }
}), Ui = g, Ln = A({
  name: "RunButton",
  props: {
    onClick: {
      type: Function,
      default: null
    }
  },
  setup(e) {
    const { sandpack: n } = z();
    return () => s(Se, {
      class: Ui.toString(),
      onClick: (t) => {
        n && n.runSandpack && n.runSandpack(), e.onClick && e.onClick(t);
      }
    }, {
      default: () => [s(kn, null, null), s("span", null, [w("Run")])]
    });
  }
}), yt = g, Ee = A({
  name: "SandpackStack",
  setup(e, { slots: n, attrs: t }) {
    const o = R();
    return () => s("div", {
      class: o("stack", [yt, (t == null ? void 0 : t.class) || ""]),
      style: (t == null ? void 0 : t.style) || {}
    }, [n.default ? n.default() : null]);
  }
}), zi = g, Nn = A({
  name: "SandpackLayout",
  setup(e, { slots: n, attrs: t }) {
    const o = T(), { sandpack: l } = z(), c = R();
    return H(o, () => {
      l && o.value && (l.lazyAnchorRef = o);
    }), () => s("div", {
      ref: o,
      class: c("layout", [zi, (t == null ? void 0 : t.class) || ""])
    }, [n.default ? n.default() : null]);
  }
}), Ji = () => {
  const { sandpack: e } = z();
  return k(() => {
    var t;
    return (t = e.error) != null ? t : {
      message: ""
    };
  });
}, Yi = (e) => {
  const n = e.replace("[sandpack-client]: ", "");
  if (/process.exit/.test(n)) {
    const t = n.match(/process.exit\((\d+)\)/);
    return t ? Number(t[1]) === 0 ? "Server is not running, would you like to start it again?" : `Server has crashed with status code ${t[1]}, would you like to restart the server?` : n;
  }
  return n;
}, gt = A({
  name: "ErrorOverlay",
  props: {
    clientId: {
      type: String,
      required: !1,
      default: ""
    }
  },
  setup(e, { slots: n, attrs: t }) {
    const o = Ji(), { restart: l } = Ye(), { sandpack: c, dispatch: d } = z(), a = R(), f = k(() => {
      var h, u;
      return (u = (h = o.value) == null ? void 0 : h.message) == null ? void 0 : u.startsWith("[sandpack-client]");
    }), y = k(() => {
      var h, u;
      return (u = (h = o.value) == null ? void 0 : h.message) == null ? void 0 : u.includes("NPM_REGISTRY_UNAUTHENTICATED_REQUEST");
    }), m = () => {
      c != null && c.teamId && d({
        type: "sign-in",
        teamId: c.teamId
      });
    };
    return () => {
      var h, u, C;
      return s(Q, null, [y.value ? s("div", me(e, {
        class: a("overlay", [a("error"), Fe, ut, (t == null ? void 0 : t.class) || ""])
      }), [s("p", {
        class: a("error-message", [ge])
      }, [s("strong", null, [w("Unable to fetch required dependency.")])]), s("div", {
        class: a("error-message", [ge])
      }, [s("p", null, [w("Authentication required. Please sign in to your account (make sure to allow pop-ups to this page) and try again. If the issue persists, contact"), " ", s("a", {
        href: "mailto:hello@codesandbox.io?subject=Sandpack Timeout Error"
      }, [w("support")]), " ", w("for further assistance.")])]), s("div", null, [s("button", {
        class: a("button", [le, Le, Ce]),
        onClick: m
      }, [s(Pi, null, null), s("span", null, [w("Sign in")])])])]) : s(Q, null, [f.value && ((h = o.value) != null && h.message) ? s("div", me({
        class: a("overlay", [a("error"), Fe, ut, (t == null ? void 0 : t.class) || ""])
      }, e), [s("div", {
        class: a("error-message", [ge])
      }, [s("p", {
        class: a("error-title", [g])
      }, [w("Couldn't connect to server")]), s("p", null, [Yi((u = o.value) == null ? void 0 : u.message)]), s("div", null, [s("button", {
        class: a("button", [a("icon-standalone"), le, Le, Ce]),
        onClick: () => {
          l(), c && c.runSandpack();
        },
        title: "Restart script",
        type: "button"
      }, [s(We, null, null), w(" "), s("span", null, [w("Restart")])])])])]) : !o.value.message && !n.default ? null : s("div", {
        class: a("overlay", [a("error"), Fe, xn(), (t == null ? void 0 : t.class) || ""]),
        translate: "no"
      }, [s("p", {
        class: a("error-message", [ge])
      }, [s("strong", null, [w("Something went wrong")])]), s("p", {
        class: a("error-message", [ge()])
      }, [((C = o.value) == null ? void 0 : C.message) || (n.default ? n.default() : null)])])])]);
    };
  }
}), Zi = (e) => ls.compressToBase64(JSON.stringify(e)).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, ""), Kt = "https://codesandbox.io/api/v1/sandboxes/define", Wi = (e, n) => {
  const t = Object.keys(e).reduce((o, l) => {
    const c = l.replace("/", ""), d = {
      content: e[l].code,
      isBinary: !1
    };
    return O(x({}, o), {
      [c]: d
    });
  }, {});
  return Zi(x({
    files: t
  }, n ? {
    template: n
  } : null));
}, Gi = A({
  name: "UnstyledOpenInCodeSandboxButton",
  setup(e, { slots: n }) {
    let t;
    const { sandpack: o } = z(), l = T(), c = T({});
    return H([() => o.activeFile, () => o.environment, () => o.files], () => {
      t && clearTimeout(t), t = setTimeout(() => {
        const d = Wi(o.files, o.environment), a = new URLSearchParams({
          parameters: d,
          query: new URLSearchParams({
            file: o.activeFile,
            utm_medium: "sandpack"
          }).toString()
        });
        c.value = a;
      }, 600);
    }, {
      deep: !0,
      immediate: !0
    }), () => {
      var d, a, f, y, m;
      return ((y = (f = (a = (d = c.value) == null ? void 0 : d.get) == null ? void 0 : a.call(d, "parameters")) == null ? void 0 : f.length) != null ? y : 0) > 1500 ? s("button", {
        onClick: () => {
          var h;
          return (h = l.value) == null ? void 0 : h.submit();
        },
        title: "Open in CodeSandbox",
        type: "button"
      }, [s("form", {
        ref: l,
        action: Kt,
        method: "POST",
        style: {
          visibility: "hidden"
        },
        target: "_blank"
      }, [s("input", {
        name: "environment",
        type: "hidden",
        value: o.environment === "node" ? "server" : o.environment
      }, null), Array.from(c.value, ([h, u]) => s("input", {
        key: h,
        name: h,
        type: "hidden",
        value: u
      }, null))]), n.default ? n.default() : null]) : s("a", {
        href: `${Kt}?${(m = c.value) == null ? void 0 : m.toString()}&environment=${o.environment === "node" ? "server" : o.environment}`,
        rel: "noreferrer noopener",
        target: "_blank",
        title: "Open in CodeSandbox"
      }, [n.default ? n.default() : null]);
    };
  }
}), On = A({
  name: "OpenInCodeSandboxButton",
  setup() {
    const e = R();
    return () => s(Gi, {
      class: e("button", [e("icon-standalone"), le, Le, Ce])
    }, {
      default: () => [s(ji, null, null), s("span", null, [w("Open Sandbox")])]
    });
  }
}), Ki = g, Xi = g;
vn({
  "0%": {
    transform: "rotateX(-25.5deg) rotateY(45deg)"
  },
  "100%": {
    transform: "rotateX(-25.5deg) rotateY(405deg)"
  }
});
const Qi = g, Mn = A({
  name: "Loading",
  props: {
    showOpenInCodeSandbox: {
      type: Boolean,
      required: !0,
      default: !1
    }
  },
  setup(e, { attrs: n }) {
    const t = R();
    return () => s("div", {
      class: t("cube-wrapper", [Xi, (n == null ? void 0 : n.class) || ""]),
      title: "Open in CodeSandbox"
    }, [e.showOpenInCodeSandbox && s(On, null, null), s("div", {
      class: t("cube", [Ki])
    }, [s("div", {
      class: t("sides", [Qi])
    }, [s("div", {
      class: "top"
    }, null), s("div", {
      class: "right"
    }, null), s("div", {
      class: "bottom"
    }, null), s("div", {
      class: "left"
    }, null), s("div", {
      class: "front"
    }, null), s("div", {
      class: "back"
    }, null)])])]);
  }
});
function eo(e) {
  return typeof e == "function" || Object.prototype.toString.call(e) === "[object Object]" && !ts(e);
}
const Fn = A({
  name: "StdoutList",
  props: {
    data: {
      type: Array,
      required: !1,
      default() {
        return [];
      }
    }
  },
  setup(e) {
    const n = R();
    return () => s(Q, null, [e.data.map(({ data: t, id: o }) => s("div", {
      key: o,
      class: n("console-item", [to])
    }, [s(as, {
      linkify: !0
    }, eo(t) ? t : {
      default: () => [t]
    })]))]);
  }
}), to = g, no = (e, n) => {
  var t;
  switch (e.state) {
    case "downloading_manifest":
      return "[1/3] Downloading manifest";
    case "downloaded_module":
      return `[2/3] Downloaded ${e.name} (${n - e.totalPending}/${n})`;
    case "starting_command":
      return "[3/3] Starting command";
    case "command_running":
      return `[3/3] Running "${(t = e.command) == null ? void 0 : t.trim()}"`;
  }
}, jn = (e) => {
  let n, t;
  const o = T(!1), l = T(), c = T(null), { listen: d } = z();
  return H([l, o, () => e == null ? void 0 : e.timeout, () => e == null ? void 0 : e.clientId], () => {
    n && n(), n = d((a) => {
      a.type === "start" && a.firstLoad && (o.value = !1), e != null && e.timeout && (t = setTimeout(() => {
        c.value = null;
      }, e == null ? void 0 : e.timeout)), a.type === "shell/progress" && !o.value && (!l.value && a.data.state === "downloaded_module" && (l.value = a.data.totalPending), l.value !== void 0 && (c.value = no(a.data, l.value))), a.type === "done" && a.compilatonError === !1 && (o.value = !0, c.value = null, clearTimeout(t));
    }, e == null ? void 0 : e.clientId);
  }, {
    immediate: !0
  }), ae(() => {
    t && clearTimeout(t), n && n();
  }), se(() => {
    t && clearTimeout(t), n && n();
  }), c;
}, so = 400 * 2, Rn = (e) => {
  let n;
  const t = T([]), { listen: o } = z(), l = k(() => {
    var d;
    return (d = e == null ? void 0 : e.maxMessageCount) != null ? d : so;
  }), c = k(() => {
    var d;
    return (d = e == null ? void 0 : e.resetOnPreviewRestart) != null ? d : !1;
  });
  return H([l, () => e == null ? void 0 : e.clientId], () => {
    n && n(), n = o((d) => {
      if (c.value && d.type === "start")
        t.value = [];
      else if (d.type === "stdout" && d.payload.data && d.payload.data.trim()) {
        const f = [...[...V(t.value)], {
          data: d.payload.data,
          id: ze()
        }];
        for (; f.length > l.value; )
          f.shift();
        t.value = V(f);
      }
    }, e == null ? void 0 : e.clientId);
  }, {
    immediate: !0
  }), ae(() => {
    n && n();
  }), se(() => {
    n && n();
  }), {
    logs: t,
    reset: () => {
      t.value = [];
    }
  };
}, io = g, bt = A({
  name: "LoadingOverlay",
  props: {
    clientId: String,
    loading: Boolean,
    showOpenInCodeSandbox: {
      type: Boolean,
      required: !0,
      default: !1
    }
  },
  setup(e, { attrs: n }) {
    let t;
    const { sandpack: o } = z(), { restart: l } = Ye(e.clientId), c = R(), d = T(!1), a = Si(e), f = jn({
      clientId: e == null ? void 0 : e.clientId
    }), { logs: y } = Rn({
      clientId: e == null ? void 0 : e.clientId
    }), m = k(() => a.value !== "HIDDEN"), h = k(() => a.value === "TIMEOUT"), u = k(() => a.value === "LOADING"), C = k(() => u.value || a.value === "PRE_FADING"), r = () => s("div", {
      class: c("overlay", [c("error"), Fe, xn, ut, (n == null ? void 0 : n.class) || ""])
    }, [s("div", {
      class: c(c("error-message", [ge]))
    }, [s("p", {
      class: c("error-title", [g])
    }, [w("Couldn't connect to server")]), s("div", {
      class: c("error-message", [ge])
    }, [s("p", null, [w("This means sandpack cannot connect to the runtime or your network is having some issues. Please check the network tab in your browser and try again. If the problem persists, report it via"), " ", s("a", {
      href: "mailto:hello@codesandbox.io?subject=Sandpack Timeout Error"
    }, [w("email")]), " ", w("or submit an issue on"), " ", s("a", {
      href: "https://github.com/codesandbox/sandpack/issues",
      rel: "noreferrer noopener",
      target: "_blank"
    }, [w("GitHub.")])])]), s("p", {
      class: c("error-message", [ge()])
    }, [w("ENV: "), o.environment, s("br", null, null), w("ERROR: TIME_OUT")]), s("div", null, [s("button", {
      class: c("button", [c("icon-standalone"), le, Le, Ce]),
      onClick: () => ke(this, null, function* () {
        o && (yield o.runSandpack()), l();
      }),
      title: "Restart script",
      type: "button"
    }, [s(We, null, null), s("span", null, [w("Try again")])])])])]);
    return H(f, () => {
      var P;
      t && clearTimeout(t), (P = f.value) != null && P.includes("Running") && (t = setTimeout(() => {
        d.value = !0;
      }, 3e3));
    }, {
      immediate: !0
    }), ae(() => {
      t && clearTimeout(t);
    }), se(() => {
      t && clearTimeout(t);
    }), () => s(Q, null, [m.value ? h.value ? r() : u.value ? s(Q, null, [s("div", {
      class: c("overlay", [c("loading"), Fe, io, (n == null ? void 0 : n.class) || ""]),
      style: O(x({}, n.style || {}), {
        opacity: o && C.value && o.status && o.status !== "idle" ? 1 : 0,
        transition: `opacity ${Cn}ms ease-out`
      })
    }, [d.value && s("div", {
      class: oo.toString()
    }, [s(Fn, {
      data: y.value
    }, null)]), s(Mn, {
      showOpenInCodeSandbox: e.showOpenInCodeSandbox
    }, null)]), f.value && s("div", {
      class: lo.toString()
    }, [s("p", null, [f.value])])]) : null : null]);
  }
}), oo = g, lo = g, ao = A({
  name: "SandpackLayout",
  props: {
    clientId: {
      type: String,
      required: !1,
      default: void 0
    }
  },
  setup(e) {
    const n = jn({
      timeout: 3e3,
      clientId: e == null ? void 0 : e.clientId
    });
    return () => s(Q, null, [n.value ? s("div", {
      class: ro.toString()
    }, [s("p", null, [n.value])]) : null]);
  }
}), ro = g, In = A({
  name: "SandpackCodeEditor",
  props: {
    showTabs: {
      type: Boolean,
      required: !1,
      default: void 0
    },
    showLineNumbers: {
      type: Boolean,
      required: !1,
      default: !1
    },
    showInlineErrors: {
      type: Boolean,
      required: !1,
      default: !1
    },
    wrapContent: {
      type: Boolean,
      required: !1,
      default: !1
    },
    closableTabs: {
      type: Boolean,
      required: !1
    },
    initMode: {
      type: String,
      required: !1,
      default: void 0
    },
    extensions: {
      type: Array,
      required: !1,
      default: void 0
    },
    extensionsKeymap: {
      type: Array,
      required: !1,
      default: void 0
    },
    readOnly: {
      type: Boolean,
      required: !1,
      default: void 0
    },
    showReadOnly: {
      type: Boolean,
      required: !1,
      default: void 0
    },
    additionalLanguages: {
      type: Array,
      required: !1,
      default: void 0
    }
  },
  setup(e, { attrs: n }) {
    const { sandpack: t } = z(), { code: o, readOnly: l } = bn(), c = k(() => {
      var m, h;
      return (h = e.showTabs) != null ? h : ((m = t == null ? void 0 : t.visibleFiles) == null ? void 0 : m.length) > 1;
    }), d = k(() => !(t != null && t.autorun)), a = T(null), f = R(), y = (m, h = !0) => {
      t.updateCurrentFile(m, h);
    };
    return () => s(Ee, {
      class: f("editor", [(n == null ? void 0 : n.class) || ""]),
      style: (n == null ? void 0 : n.style) || {}
    }, {
      default: () => [c.value && s(An, {
        closableTabs: e.closableTabs
      }, null), s("div", {
        class: f("code-editor", [En])
      }, [s(Ze, {
        key: t.activeFile,
        ref: a,
        code: o.value,
        editorState: t.editorState,
        extensions: e.extensions,
        extensionsKeymap: e.extensionsKeymap,
        filePath: t.activeFile,
        initMode: e.initMode || t.initMode,
        onCodeUpdate: (m) => {
          var h;
          y(m, (h = t.autoReload) != null ? h : !0);
        },
        readOnly: e.readOnly || l.value,
        showInlineErrors: e.showInlineErrors,
        showLineNumbers: e.showLineNumbers,
        showReadOnly: e.showReadOnly,
        wrapContent: e.wrapContent,
        additionalLanguages: e.additionalLanguages
      }, null), d.value && (!t.autoReload || t.status === "idle") ? s(Ln, null, null) : null])]
    });
  }
}), co = g, Dn = A({
  name: "File",
  props: {
    path: {
      type: String
    },
    selectFile: {
      type: Function,
      required: !1,
      default: void 0
    },
    active: {
      type: Boolean,
      required: !1,
      default: void 0
    },
    onClick: {
      type: Function,
      required: !1,
      default: void 0
    },
    depth: {
      type: Number
    },
    isDirOpen: {
      type: Boolean,
      required: !1,
      default: void 0
    }
  },
  setup(e) {
    const n = R(), t = k(() => e.path.split("/").filter(Boolean).pop()), o = (l) => {
      e.selectFile && e.selectFile(e.path), e.onClick && e.onClick(l);
    };
    return () => s("button", {
      class: n("button", [n("explorer"), le, co]),
      "data-active": e.active,
      onClick: o,
      style: {
        paddingLeft: 9 * e.depth + "px"
      },
      type: "button"
    }, [e.selectFile ? s(Di, null, null) : e.isDirOpen ? s(Ri, null, null) : s(Ii, null, null), s("span", null, [t.value])]);
  }
}), uo = ({ autoHiddenFiles: e, visibleFiles: n = [], files: t, prefixedPath: o }) => {
  const l = n.length > 0, c = e && !l, d = e && !!l, a = Object.keys(t).filter((m) => {
    var u;
    const h = m.startsWith(o);
    return d ? h && n.includes(m) : c ? h && !((u = t[m]) != null && u.hidden) : h;
  }).map((m) => m.substring(o.length)), f = Array.from(new Set(a.filter((m) => m.includes("/")).map((m) => `${o}${m.split("/")[0]}/`))), y = a.filter((m) => !m.includes("/")).map((m) => `${o}${m}`);
  return {
    directories: f,
    modules: y
  };
}, Bn = A({
  name: "ModuleList",
  props: {
    prefixedPath: {
      type: String
    },
    files: {
      type: Object
    },
    selectFile: {
      type: Function
    },
    activeFile: {
      type: String
    },
    depth: {
      type: Number,
      required: !1,
      default: 0
    },
    autoHiddenFiles: {
      type: Boolean,
      required: !1,
      default: !1
    },
    visibleFiles: {
      type: Array,
      required: !1,
      default: void 0
    },
    initialCollapsedFolder: {
      type: Array,
      required: !1,
      default() {
        return [];
      }
    }
  },
  setup(e) {
    const n = k(() => uo({
      visibleFiles: e.visibleFiles,
      autoHiddenFiles: e.autoHiddenFiles,
      prefixedPath: e.prefixedPath,
      files: e.files
    }));
    return () => s("div", null, [Array.from(n.value.directories).map((t) => s(fo, {
      key: t,
      activeFile: e.activeFile,
      autoHiddenFiles: e.autoHiddenFiles,
      depth: e.depth,
      files: e.files,
      initialCollapsedFolder: e.initialCollapsedFolder,
      prefixedPath: t,
      selectFile: e.selectFile,
      visibleFiles: e.visibleFiles
    }, null)), n.value.modules.map((t) => s(Dn, {
      key: t,
      active: e.activeFile === t,
      depth: e.depth,
      path: t,
      selectFile: e.selectFile
    }, null))]);
  }
}), fo = A({
  name: "Directory",
  props: {
    prefixedPath: {
      type: String
    },
    files: {
      type: Object
    },
    selectFile: {
      type: Function
    },
    activeFile: {
      type: String
    },
    depth: {
      type: Number,
      required: !1,
      default: 0
    },
    visibleFiles: {
      type: Array,
      required: !1,
      default: void 0
    },
    initialCollapsedFolder: {
      type: Array,
      required: !1,
      default() {
        return [];
      }
    },
    autoHiddenFiles: {
      type: Boolean,
      required: !1,
      default: !1
    }
  },
  setup(e) {
    var o;
    const n = T(!((o = e.initialCollapsedFolder) != null && o.includes(e.prefixedPath || ""))), t = () => {
      n.value = !n.value;
    };
    return () => s("div", {
      key: e.prefixedPath
    }, [s(Dn, {
      depth: e.depth,
      isDirOpen: n.value,
      onClick: t,
      path: e.prefixedPath + "/"
    }, null), n.value ? s(Bn, {
      autoHiddenFiles: e.autoHiddenFiles,
      activeFile: e.activeFile,
      depth: e.depth + 1,
      files: e.files,
      initialCollapsedFolder: e.initialCollapsedFolder,
      prefixedPath: e.prefixedPath,
      selectFile: e.selectFile,
      visibleFiles: e.visibleFiles
    }, null) : null]);
  }
}), po = g, mo = A({
  name: "SandpackFileExplorer",
  props: {
    autoHiddenFiles: {
      type: Boolean,
      required: !1,
      default: !1
    },
    initialCollapsedFolder: {
      type: Array,
      required: !1,
      default() {
        return [];
      }
    }
  },
  setup(e, { attrs: n }) {
    let t;
    const { sandpack: o, listen: l } = z(), c = R(), d = k(() => Object.keys(o.files).sort().reduce((a, f) => (a[f] = o.files[f], a), {}));
    return H(() => o.status, () => {
      if (o.status === "running")
        return t && t(), t = l((a) => {
          a.type === "fs/change" && o.updateFile(a.path, a.content, !1), a.type === "fs/remove" && o.deleteFile(a.path, !1);
        }), t;
    }, {
      immediate: !0
    }), ae(() => {
      t && t();
    }), se(() => {
      t && t();
    }), () => {
      var a;
      return s("div", {
        class: c("file-explorer", [yt, (n == null ? void 0 : n.class) || ""])
      }, [s("div", {
        class: c("file-explorer-list", [po])
      }, [s(Bn, {
        activeFile: o.activeFile,
        autoHiddenFiles: e.autoHiddenFiles,
        files: d.value,
        initialCollapsedFolder: (a = e == null ? void 0 : e.initialCollapsedFolder) != null ? a : [],
        prefixedPath: "/",
        selectFile: o.openFile,
        visibleFiles: o.visibleFilesFromProps
      }, null)])]);
    };
  }
}), Pn = A({
  name: "SandpackCodeViewer",
  props: {
    showTabs: {
      type: Boolean,
      required: !1,
      default: void 0
    },
    showLineNumbers: {
      type: Boolean,
      required: !1,
      default: void 0
    },
    decorators: {
      type: Object,
      required: !1,
      default: void 0
    },
    code: {
      type: String,
      required: !1,
      default: void 0
    },
    initMode: {
      type: String,
      required: !1,
      default: void 0
    },
    wrapContent: {
      type: String,
      required: !1,
      default: void 0
    },
    additionalLanguages: {
      type: Array,
      required: !1,
      default: void 0
    }
  },
  setup(e) {
    const { sandpack: n } = z(), { code: t } = bn(), o = R(), l = T(), c = k(() => {
      var d, a;
      return (a = e.showTabs) != null ? a : ((d = n == null ? void 0 : n.visibleFiles) == null ? void 0 : d.length) > 1;
    });
    return () => s(Ee, {
      class: o("editor-viewer")
    }, {
      default: () => {
        var d;
        return [c.value ? s(An, null, null) : null, s("div", {
          class: o("code-editor", [En])
        }, [s(Ze, {
          ref: l,
          code: (d = e.code) != null ? d : t.value,
          additionalLanguages: e.additionalLanguages,
          decorators: e.decorators,
          filePath: n.activeFile,
          initMode: e.initMode || n.initMode,
          showLineNumbers: e.showLineNumbers,
          showReadOnly: !1,
          wrapContent: e.wrapContent,
          readOnly: !0
        }, null)]), n.status === "idle" ? s(Ln, null, null) : null];
      }
    });
  }
}), vo = g, ho = g, yo = (e) => {
  const n = e.match(/(https?:\/\/.*?)\//);
  return n && n[1] ? [n[1], e.replace(n[1], "")] : [e, "/"];
}, go = A({
  name: "Navigator",
  props: {
    clientId: {
      type: String,
      required: !0
    },
    onURLChange: {
      type: Function,
      required: !1,
      default: void 0
    },
    startRoute: {
      type: String,
      required: !1,
      default: void 0
    }
  },
  setup(e, { attrs: n }) {
    var J, W;
    let t;
    const o = T(""), { sandpack: l, dispatch: c, listen: d } = z(), a = T((W = (J = e == null ? void 0 : e.startRoute) != null ? J : l.startRoute) != null ? W : "/"), f = T(!1), y = T(!1), m = R();
    H(() => e.clientId, () => {
      t && t(), t = d((L) => {
        if (L.type === "urlchange") {
          const { url: I, back: K, forward: re } = L, [F, S] = yo(I);
          o.value = F, a.value = S, f.value = K, y.value = re;
        }
      }, e.clientId);
    }, {
      immediate: !0
    });
    const h = (L) => {
      const I = L.target.value.startsWith("/") ? L.target.value : `/${L.target.value}`;
      a.value = I;
    }, u = (L) => {
      var I;
      L.code === "Enter" && (L.preventDefault(), L.stopPropagation(), typeof e.onURLChange == "function" && e.onURLChange(o.value + ((I = L.currentTarget) == null ? void 0 : I.value)));
    }, C = () => {
      c({
        type: "refresh"
      });
    }, r = () => {
      c({
        type: "urlback"
      });
    }, P = () => {
      c({
        type: "urlforward"
      });
    }, D = m("button", [m("icon"), le, Li, g]);
    return ae(() => {
      t && t();
    }), se(() => {
      t && t();
    }), () => s("div", {
      class: m("navigator", [vo, (n == null ? void 0 : n.class) || ""])
    }, [s("button", {
      "aria-label": "Go back one page",
      class: D,
      disabled: !f.value,
      onClick: r,
      type: "button"
    }, [s(Oi, null, null)]), s("button", {
      "aria-label": "Go forward one page",
      class: D,
      disabled: !y.value,
      onClick: P,
      type: "button"
    }, [s(Mi, null, null)]), s("button", {
      "aria-label": "Refresh page",
      class: D,
      onClick: C,
      type: "button"
    }, [s(Tn, null, null)]), s("input", {
      "aria-label": "Current Sandpack URL",
      class: m("input", [ho]),
      name: "Current Sandpack URL",
      onChange: h,
      onKeydown: u,
      type: "text",
      value: a.value
    }, null)]);
  }
}), bo = g, Co = g, wo = g, qn = A({
  name: "SandpackPreview",
  props: {
    showNavigator: {
      type: Boolean,
      required: !1,
      default: !1
    },
    showRefreshButton: {
      type: Boolean,
      required: !1,
      default: !0
    },
    showOpenInCodeSandbox: {
      type: Boolean,
      required: !1,
      default: !0
    },
    showSandpackErrorOverlay: {
      type: Boolean,
      required: !1,
      default: !0
    },
    showOpenNewtab: {
      type: Boolean,
      required: !1,
      default: !0
    },
    showRestartButton: {
      type: Boolean,
      required: !1,
      default: !0
    },
    actionsChildren: {
      type: Object,
      required: !1,
      default: null
    },
    startRoute: {
      type: String,
      required: !1,
      default: "/"
    }
  },
  setup(e, { slots: n, attrs: t, expose: o }) {
    const { sandpack: l, listen: c, iframe: d, getClient: a, clientId: f, dispatch: y } = vt({
      startRoute: e == null ? void 0 : e.startRoute
    }), m = T(null);
    let h;
    const u = R(), { refresh: C } = xi(f.value), { restart: r } = Ye(f.value);
    xe(() => {
      Ie(() => {
        h && h(), h = c((D) => {
          D.type === "resize" && (m.value = D.height);
        });
      });
    }), ae(() => {
      h && h();
    }), se(() => {
      h && h();
    }), o({
      clientId: f.value,
      getClient: a
    });
    const P = (D) => {
      d.value && (d.value.src = D);
    };
    return () => s(Ee, me(e, {
      class: u("preview", [(t == null ? void 0 : t.class) || ""])
    }), {
      default: () => [e.showNavigator && s(go, {
        clientId: f.value,
        startRoute: e == null ? void 0 : e.startRoute,
        onURLChange: P
      }, null), s("div", {
        class: u("preview-container", [bo])
      }, [s("iframe", {
        ref: d,
        class: u("preview-iframe", [Co]),
        style: {
          height: m.value ? `${m.value}px` : void 0
        },
        title: "Sandpack Preview"
      }, null), s("div", {
        class: u("preview-actions", [wo])
      }, [n.actionsChildren ? n.actionsChildren() : e.actionsChildren ? e.actionsChildren : null, e.showRestartButton && l.environment === "node" && s(Se, {
        onClick: r
      }, {
        default: () => [s(We, null, null)]
      }), !e.showNavigator && e.showRefreshButton && l.status === "running" && s(Se, {
        onClick: C
      }, {
        default: () => [s(Tn, null, null)]
      }), l.teamId && s("button", {
        class: u("button", [u("icon-standalone"), le, Le, Ce]),
        onClick: () => y({
          type: "sign-out"
        }),
        title: "Sign out",
        type: "button"
      }, [s(qi, null, null)]), e.showOpenInCodeSandbox && s(On, null, null)]), s(bt, {
        clientId: f.value,
        showOpenInCodeSandbox: e.showOpenInCodeSandbox
      }, null), e.showSandpackErrorOverlay && s(gt, {
        clientId: f.value
      }, null), n.default ? n.default() : null])]
    });
  }
}), So = g, xo = A({
  name: "SandpackTranspiledCode",
  props: {
    showTabs: {
      type: Boolean,
      required: !1,
      default: void 0
    },
    showLineNumbers: {
      type: Boolean,
      required: !1,
      default: void 0
    },
    decorators: {
      type: Array,
      required: !1,
      default: void 0
    },
    code: {
      type: String,
      required: !1,
      default: void 0
    },
    wrapContent: {
      type: Boolean,
      required: !1,
      default: void 0
    },
    initMode: {
      type: String,
      required: !1,
      default: void 0
    }
  },
  setup(e, { attrs: n }) {
    const { sandpack: t } = z(), o = R(), l = T(null), c = k(() => t.bundlerState), d = k(() => {
      var f, y, m;
      const a = (f = c.value) == null ? void 0 : f.transpiledModules[t.activeFile + ":"];
      return (m = (y = a == null ? void 0 : a.source) == null ? void 0 : y.compiledCode) != null ? m : "";
    });
    return xe(() => {
      const a = l.value;
      a && t.registerBundler(a, "hidden");
    }), se(() => {
      t.unregisterBundler("hidden");
    }), () => {
      var a;
      return s("div", {
        class: o("transpiled-code", [yt, So, (n == null ? void 0 : n.class) || ""])
      }, [s(Pn, me(e, {
        code: (a = d.value) != null ? a : "",
        initMode: t.initMode
      }), null), s("iframe", {
        ref: l,
        style: {
          display: "none"
        },
        title: "transpiled sandpack code"
      }, null), s(gt, {
        clientId: "hidden"
      }, null), s(bt, {
        clientId: "hidden",
        showOpenInCodeSandbox: !1
      }, null)]);
    };
  }
}), Eo = ["SyntaxError: ", "Error in sandbox:"], ko = {
  id: "random",
  method: "clear",
  data: ["Console was cleared"]
}, Xt = "@t", Qt = "#@t", en = "@r", tn = 1e4, $n = 2, dt = 400, nn = dt * 2, tt = function() {
  return (0, eval)("this");
}(), To = typeof ArrayBuffer == "function", Ao = typeof Map == "function", Lo = typeof Set == "function";
var No = ((e) => (e[e.infinity = 0] = "infinity", e[e.minusInfinity = 1] = "minusInfinity", e[e.minusZero = 2] = "minusZero", e))(No || {});
const sn = {
  Arithmetic: (e) => e === 0 ? 1 / 0 : e === 1 ? -1 / 0 : e === 2 ? -0 : e,
  HTMLElement: (e) => {
    const n = document.implementation.createHTMLDocument("sandbox");
    try {
      const t = n.createElement(e.tagName);
      t.innerHTML = e.innerHTML;
      for (const o of Object.keys(e.attributes))
        try {
          t.setAttribute(o, e.attributes[o]);
        } catch (l) {
        }
      return t;
    } catch (t) {
      return e;
    }
  },
  Function: (e) => {
    const n = () => {
    };
    return Object.defineProperty(n, "toString", {
      value: () => `function ${e.name}() {${e.body}}`
    }), n;
  },
  "[[NaN]]": () => NaN,
  "[[undefined]]": () => {
  },
  "[[Date]]": (e) => {
    const n = /* @__PURE__ */ new Date();
    return n.setTime(e), n;
  },
  "[[RegExp]]": (e) => new RegExp(e.src, e.flags),
  "[[Error]]": (e) => {
    const n = tt[e.name] || Error, t = new n(e.message);
    return t.stack = e.stack, t;
  },
  "[[ArrayBuffer]]": (e) => {
    if (To) {
      const n = new ArrayBuffer(e.length);
      return new Int8Array(n).set(e), n;
    }
    return e;
  },
  "[[TypedArray]]": (e) => typeof tt[e.ctorName] == "function" ? new tt[e.ctorName](e.arr) : e.arr,
  "[[Map]]": (e) => {
    if (Ao) {
      const t = /* @__PURE__ */ new Map();
      for (let o = 0; o < e.length; o += 2)
        t.set(e[o], e[o + 1]);
      return t;
    }
    const n = [];
    for (let t = 0; t < e.length; t += 2)
      n.push([e[i], e[i + 1]]);
    return n;
  },
  "[[Set]]": (e) => {
    if (Lo) {
      const n = /* @__PURE__ */ new Set();
      for (let t = 0; t < e.length; t++)
        n.add(e[t]);
      return n;
    }
    return e;
  }
}, ft = (e) => {
  var n;
  if (typeof e == "string" || typeof e == "number" || e === null)
    return e;
  if (Array.isArray(e))
    return e.map(ft);
  if (typeof e == "object" && Xt in e) {
    const t = e[Xt], o = sn[t];
    return o(e.data);
  } else if (typeof e == "object" && Qt in e) {
    const t = e[Qt], o = sn[t];
    return o(e.data);
  } else if (typeof e == "object" && ((n = e.constructor) == null ? void 0 : n.name) === "NodeList") {
    const t = {};
    return Object.entries(e).forEach(([o, l]) => {
      t[o] = ft(l);
    }), t;
  }
  return e;
}, Oo = (e, n, t) => `[${e.reduce((l, c, d) => `${l}${d ? ", " : ""}${Re(c, n, t)}`, "")}]`, Mo = (e, n, t) => {
  const o = e.constructor.name !== "Object" ? `${e.constructor.name} ` : "";
  if (t > $n)
    return o;
  const l = Object.entries(e), c = Object.entries(e).reduce((d, [a, f], y) => {
    const m = y === 0 ? "" : ", ", h = l.length > 10 ? `
  ` : "", u = Re(f, n, t);
    return y === dt ? d + h + "..." : y > dt ? d : d + `${m}${h}${a}: ` + u;
  }, "");
  return `${o}{ ${c}${l.length > 10 ? `
` : " "}}`;
}, Re = (e, n, t = 0) => {
  var o;
  try {
    const l = ft(e);
    if (Array.isArray(l))
      return Oo(l, n, t + 1);
    switch (typeof l) {
      case "string":
        return `"${l}"`.slice(0, tn);
      case "number":
      case "function":
      case "symbol":
        return l.toString();
      case "boolean":
        return String(l);
      case "undefined":
        return "undefined";
      case "object":
      default:
        if (l instanceof RegExp || l instanceof Error || l instanceof Date)
          return l.toString();
        if (l === null)
          return String(null);
        if (l instanceof HTMLElement)
          return l.outerHTML.slice(0, tn);
        if (Object.entries(l).length === 0)
          return "{}";
        if (en in l) {
          if (t > $n)
            return "Unable to print information";
          const c = n[l[en]];
          return Re(c, n, t + 1);
        }
        if (((o = l.constructor) == null ? void 0 : o.name) === "NodeList") {
          const { length: c } = l, d = new Array(c).fill(null).map((a, f) => Re(l[f], n));
          return `NodeList(${l.length})[${d}]`;
        }
        return Mo(l, n, t + 1);
    }
  } catch (l) {
    return "Unable to print information";
  }
}, Fo = A({
  name: "ConsoleList",
  props: {
    data: {
      type: Array,
      required: !1,
      default() {
        return [];
      }
    }
  },
  setup(e) {
    const n = R();
    return () => s(Q, null, [e.data.map(({ data: t, id: o, method: l }, c, d) => t && Array.isArray(t) ? s(Q, {
      key: o
    }, [t.map((a, f) => {
      const y = d.slice(c, d.length);
      return s("div", {
        key: `${o}-${f}`,
        class: n("console-item", [jo()])
      }, [s(Ze, {
        code: l === "clear" ? a : Re(a, y),
        fileType: "js",
        initMode: "user-visible",
        showReadOnly: !1,
        readOnly: !0,
        wrapContent: !0
      }, null)]);
    })]) : null)]);
  }
}), jo = g, Ro = g, on = g, Io = A({
  name: "Header",
  props: {
    currentTab: {
      type: String,
      required: !0
    },
    setCurrentTab: {
      type: Function,
      required: !0
    },
    node: {
      type: Boolean,
      required: !0
    }
  },
  setup(e) {
    const n = R(), t = n("console-header-button", [le, Ce, g]);
    return () => s("div", {
      class: n("console-header", [Ro, on])
    }, [s("p", {
      class: n("console-header-title", [g])
    }, [s(ht, null, null), s("span", null, [w("Terminal")]), e.node && s("div", {
      class: n("console-header-actions", [on])
    }, [s("button", {
      class: t,
      "data-active": e.currentTab === "server",
      onClick: () => e.setCurrentTab("server"),
      type: "button"
    }, [w("Server")]), s("button", {
      class: t,
      "data-active": e.currentTab === "client",
      onClick: () => e.setCurrentTab("client"),
      type: "button"
    }, [w("Client")])])])]);
  }
}), Do = (e) => {
  let n;
  const t = T([]), { listen: o } = z(), l = k(() => {
    var f;
    return (f = e == null ? void 0 : e.resetOnPreviewRestart) != null ? f : !1;
  }), c = k(() => {
    var f;
    return (f = e == null ? void 0 : e.showSyntaxError) != null ? f : !1;
  }), d = k(() => {
    var f;
    return (f = e == null ? void 0 : e.maxMessageCount) != null ? f : nn;
  }), a = () => {
    n && n(), n = o((f) => {
      if (l.value && f.type === "start")
        t.value = [];
      else if (f.type === "console" && f.codesandbox) {
        const y = Array.isArray(f.log) ? f.log : [f.log];
        if (y.find(({ method: u }) => u === "clear")) {
          t.value = [ko];
          return;
        }
        const m = c.value ? y : y.filter((u) => {
          var r, P, D;
          return ((D = (P = (r = u == null ? void 0 : u.data) == null ? void 0 : r.filter) == null ? void 0 : P.call(r, (J) => typeof J != "string" ? !0 : Eo.filter((L) => J.startsWith(L)).length === 0)) != null ? D : []).length > 0;
        });
        if (!m)
          return;
        const h = [...t.value, ...m].filter((u, C, r) => C === r.findIndex((P) => P.id === u.id));
        for (; h.length > nn; )
          h.shift();
        t.value = h;
      }
    }, e == null ? void 0 : e.clientId);
  };
  return H([d, c, l, () => e == null ? void 0 : e.clientId], () => {
    a();
  }, {
    immediate: !0
  }), ae(() => {
    n && n();
  }), se(() => {
    n && n();
  }), {
    logs: t,
    reset: () => {
      t.value = [];
    }
  };
}, ln = A({
  name: "SandpackConsole",
  props: {
    onLogsChange: {
      type: Function,
      required: !1,
      default: void 0
    },
    showSetupProgress: {
      type: Boolean,
      required: !1,
      default: !1
    },
    showRestartButton: {
      type: Boolean,
      required: !1,
      default: !0
    },
    showResetConsoleButton: {
      type: Boolean,
      required: !1,
      default: !0
    },
    resetOnPreviewRestart: {
      type: Boolean,
      required: !1,
      default: !1
    },
    showHeader: {
      type: Boolean,
      required: !1,
      default: !0
    },
    showSyntaxError: {
      type: Boolean,
      required: !1,
      default: !1
    },
    maxMessageCount: {
      type: Number,
      required: !1,
      default: void 0
    },
    standalone: {
      type: Boolean,
      required: !1,
      default: !1
    },
    actionsChildren: {
      type: Object,
      required: !1,
      default: null
    }
  },
  setup(e, { slots: n, attrs: t, expose: o }) {
    const { sandpack: { environment: l } } = z(), c = T(), { iframe: d, clientId: a } = vt(), { restart: f } = Ye(), y = R(), m = T(l === "node" ? "server" : "client"), h = k(() => e.standalone ? a.value : void 0), { logs: u, reset: C } = Do({
      maxMessageCount: e.maxMessageCount,
      showSyntaxError: e.showSyntaxError,
      resetOnPreviewRestart: e.resetOnPreviewRestart,
      clientId: h.value
    }), { logs: r, reset: P } = Rn({
      maxMessageCount: e.maxMessageCount,
      resetOnPreviewRestart: e.resetOnPreviewRestart,
      clientId: h.value
    }), D = k(() => m.value === "server"), J = k(() => l === "node");
    return o({
      reset() {
        C(), P();
      }
    }), H([() => e.onLogsChange, u, r, m], () => {
      e.onLogsChange && e.onLogsChange(V(u.value)), Ie(() => {
        setTimeout(() => {
          c.value && (c.value.scrollTop = c.value.scrollHeight);
        });
      });
    }), () => s(Ee, me(e, {
      class: y("console", [g, (t == null ? void 0 : t.class) || ""])
    }), {
      default: () => [e.showHeader && J.value && s(Io, {
        currentTab: m.value,
        node: J.value,
        setCurrentTab: (W) => {
          m.value = W;
        }
      }, null), s("div", {
        ref: c,
        class: y("console-list", [g])
      }, [D.value ? s(Fn, {
        data: r.value
      }, null) : s(Fo, {
        data: u.value
      }, null)]), s("div", {
        class: y("console-actions", [g])
      }, [n.actionsChildren ? n.actionsChildren() : e.actionsChildren ? e.actionsChildren : null, e.showRestartButton && D.value && s(Se, {
        onClick: () => {
          f(), C(), P();
        }
      }, {
        default: () => [s(We, null, null)]
      }), e.showResetConsoleButton && s(Se, {
        onClick: () => {
          m.value === "client" ? C() : P();
        }
      }, {
        default: () => [s(Fi, null, null)]
      })]), e.standalone && s(Q, null, [s(ao, {
        clientId: h.value
      }, null), s("iframe", {
        title: "console",
        ref: d
      }, null)])]
    });
  }
}), Bo = (e) => ({
  "--test-pass": e ? "#18df16" : "#15c213",
  "--test-fail": e ? "#df162b" : "#c21325",
  "--test-skip": e ? "#eace2b" : "#c2a813",
  "--test-run": e ? "#eace2b" : "#c2a813",
  "--test-title": e ? "#3fbabe" : "#256c6f"
}), Ge = g, we = Ge(), ue = Ge(), _n = Ge(), Po = Ge(), Ct = g, qo = Ct(), $o = Ct(), an = Ct(), _o = g, nt = g, Ho = A({
  name: "Header",
  props: {
    setVerbose: {
      type: Function,
      required: !0
    },
    setSuiteOnly: {
      type: Function,
      required: !0
    },
    verbose: {
      type: Boolean,
      required: !0
    },
    suiteOnly: {
      type: Boolean,
      required: !0
    },
    status: {
      type: String,
      required: !0
    },
    watchMode: {
      type: Boolean,
      required: !0
    },
    setWatchMode: {
      type: Function,
      required: !0
    },
    showSuitesOnly: {
      type: Boolean,
      required: !0
    },
    showVerboseButton: {
      type: Boolean,
      required: !0
    },
    showWatchButton: {
      type: Boolean,
      required: !0
    },
    hideTestsAndSupressLogs: {
      type: Boolean,
      required: !0
    }
  },
  setup(e) {
    const n = R(), t = n("test-header-button", [le, Ce, g]);
    return () => s("div", {
      class: n("test-header", [_o, nt])
    }, [s("div", {
      class: n("test-header-wrapper", [nt])
    }, [s("p", {
      class: n("test-header-title", [g])
    }, [s(ht, null, null), w("Tests")])]), s("div", {
      class: n("test-header-actions", [nt])
    }, [e.showSuitesOnly && s("button", {
      class: t,
      "data-active": e.suiteOnly,
      disabled: e.status === "initialising",
      onClick: e.setSuiteOnly,
      type: "button"
    }, [w("Suite only")]), e.showVerboseButton && s("button", {
      class: t,
      "data-active": e.verbose,
      disabled: e.status === "initialising" || e.hideTestsAndSupressLogs,
      onClick: e.setVerbose,
      type: "button"
    }, [w("Verbose")]), e.showWatchButton && s("button", {
      class: t,
      "data-active": e.watchMode,
      disabled: e.status === "initialising",
      onClick: e.setWatchMode,
      type: "button"
    }, [w("Watch")])])]);
  }
}), Vo = A({
  name: "RunButton",
  props: {
    onClick: {
      type: Function,
      required: !0
    }
  },
  setup(e) {
    return () => s(Se, {
      onClick: e.onClick,
      title: "Run tests"
    }, {
      default: () => [s(kn, null, null)]
    });
  }
}), Uo = g, zo = g, Jo = g, Yo = g, st = g, Hn = A({
  name: "Tests",
  props: {
    tests: {
      type: Array,
      required: !0
    }
  },
  setup(e) {
    const n = R();
    return () => s("div", {
      class: n("test", [Uo])
    }, [e.tests.map((t) => s("div", {
      key: t.name,
      class: n("test-result", [zo])
    }, [t.status === "pass" && s("span", {
      class: n("test-pass-text", [we, st])
    }, [w("")]), t.status === "fail" && s("span", {
      class: n("test-fail-text", [ue, st])
    }, [w("")]), t.status === "idle" && s("span", {
      class: n("test-idle-text", [_n, st])
    }, [w("")]), s("span", {
      class: n("test-name-text", [Jo])
    }, [t.name]), t.duration !== void 0 && s("span", {
      class: n("test-duration-text", [Yo])
    }, [w("("), t.duration, w(" ms)")])]))]);
  }
}), Ke = (e) => {
  var n, t;
  return Object.values((n = e.tests) != null ? n : {}).concat(...Object.values((t = e.describes) != null ? t : {}).map(Ke));
}, Zo = (e) => Ke(e).filter((n) => n.status === "fail"), Wo = (e) => e.map(wt).reduce((n, t) => ({
  pass: n.pass + t.pass,
  fail: n.fail + t.fail,
  skip: n.skip + t.skip,
  total: n.total + t.total
}), {
  pass: 0,
  skip: 0,
  fail: 0,
  total: 0
}), wt = (e) => Ke(e).reduce((n, t) => ({
  pass: t.status === "pass" ? n.pass + 1 : n.pass,
  fail: t.status === "fail" ? n.fail + 1 : n.fail,
  skip: t.status === "idle" || t.status === "running" ? n.skip + 1 : n.skip,
  total: n.total + 1
}), {
  pass: 0,
  fail: 0,
  skip: 0,
  total: 0
}), Go = (e) => e.filter((n) => Object.values(n.describes).length > 0 || Object.values(n.tests).length > 0).map(wt).reduce((n, t) => ({
  pass: n.pass + (t.fail === 0 ? 1 : 0),
  fail: n.fail + (t.fail > 0 ? 1 : 0),
  total: n.total + 1
}), {
  pass: 0,
  fail: 0,
  total: 0
}), Ko = (e) => Me(e, Ke).reduce((n, t) => n + (t.duration || 0), 0), Vn = (e) => Object.values((e == null ? void 0 : e.describes) || {}).length === 0 && Object.values((e == null ? void 0 : e.tests) || {}).length === 0, Pe = (e) => {
  const n = e.length - 1, t = e.slice(0, n), o = e[n];
  return [t, o];
}, Me = (e, n) => e.map(n).reduce((t, o) => t.concat(o), []), pe = (e, n, t) => rs(e, n, t), Xo = g, Qo = g, Un = A({
  name: "Describes",
  props: {
    describes: {
      type: Array,
      required: !0
    }
  },
  setup(e) {
    const n = R();
    return () => s(Q, null, [e.describes.map((t) => {
      if (Vn(t))
        return null;
      const o = Object.values((t == null ? void 0 : t.tests) || {}), l = Object.values((t == null ? void 0 : t.describes) || {});
      return s("div", {
        key: t.name,
        class: n("test-describe", [Qo])
      }, [s("div", {
        class: n("test-name", [Xo])
      }, [t.name]), s(Hn, {
        tests: o
      }, null), s(Un, {
        describes: l
      }, null)]);
    })]);
  }
}), el = g, rn = A({
  name: "FormattedError",
  props: {
    error: {
      type: Object,
      required: !0
    },
    path: {
      type: String,
      required: !1,
      default: void 0
    }
  },
  setup(e) {
    const n = R();
    return () => s("div", {
      class: n("test-error", [el]),
      innerHTML: tl(e.error, e == null ? void 0 : e.path)
    }, null);
  }
}), qe = (e) => e.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#039;"), tl = (e, n = "") => {
  let t = "";
  if (e.matcherResult ? t = `<span>${qe(e.message).replace(/(expected)/m, `<span class="${we}">$1</span>`).replace(/(received)/m, `<span class="${ue}">$1</span>`).replace(/(Difference:)/m, "<span>$1</span>").replace(/(Expected:)(.*)/m, `<span>$1</span><span class="${we}">$2</span>`).replace(/(Received:)(.*)/m, `<span>$1</span><span class="${ue}">$2</span>`).replace(/^(-.*)/gm, `<span class="${ue}">$1</span>`).replace(/^(\+.*)/gm, `<span class="${we}">$1</span>`)}</span>` : t = qe(e.message), e.mappedErrors && e.mappedErrors[0] && e.mappedErrors[0].fileName.endsWith(n) && e.mappedErrors[0]._originalScriptCode) {
    const l = e.mappedErrors[0]._originalScriptCode || [], c = Math.max(...l.map((a) => (a.lineNumber + "").length)) + 2, d = Array.from({
      length: c
    }).map(() => " ");
    t += "<br />", t += "<br />", t += "<div>", l.filter((a) => a.content.trim()).forEach((a) => {
      const f = (a.lineNumber + "").length, y = [...d];
      y.length -= f, a.highlight && (y.length -= 2);
      const m = a.content.indexOf(".to"), h = Array.from({
        length: d.length + m - (c - 1)
      }, () => " "), u = qe(a.content).replace(/(describe|test|it)(\()(&#039;|&quot;|`)(.*)(&#039;|&quot;|`)/m, `<span>$1$2$3</span><span class="${Po}">$4</span><span>$5</span>`).replace(/(expect\()(.*)(\)\..*)(to[\w\d]*)(\()(.*)(\))/m, `<span>$1</span><span class="${ue}">$2</span><span>$3</span><span style="text-decoration: underline; font-weight: 900">$4</span><span>$5</span><span class="${we}">$6</span><span>$7</span>`);
      t += `<div ${a.highlight ? 'style="font-weight:200;"' : ""}>` + (a.highlight ? `<span class="${ue}">></span> ` : "") + y.join("") + qe("" + a.lineNumber) + " | " + u + "</div>" + (a.highlight ? "<div>" + d.join("") + " | " + h.join("") + `<span class="${ue}">^</span></div>` : "");
    }), t += "</div>";
  }
  return t.replace(/(?:\r\n|\r|\n)/g, "<br />");
}, nl = g, it = g, sl = g, $e = g, il = g, ol = g, ll = g, al = g, rl = A({
  name: "Specs",
  props: {
    specs: {
      type: Array,
      required: !0
    },
    verbose: {
      type: Boolean,
      required: !0
    },
    status: {
      type: String,
      required: !0
    },
    openSpec: {
      type: Function,
      required: !0
    },
    hideTestsAndSupressLogs: {
      type: Boolean,
      required: !1,
      default: void 0
    }
  },
  setup(e) {
    const n = R();
    return () => s(Q, null, [e.specs.map((t) => {
      if (t.error)
        return s("div", {
          key: t.name,
          class: n("test-spec", [it])
        }, [s(_e, {
          class: n("test-spec-error", [$e, an])
        }, {
          default: () => [w("Error")]
        }), s(cn, {
          onClick: () => {
            e.openSpec(t.name);
          },
          path: t.name
        }, null), s(rn, {
          error: t.error,
          path: t.name
        }, null)]);
      if (Vn(t))
        return null;
      const o = Object.values(t.tests), l = Object.values(t.describes), c = wt(t);
      return s("div", {
        key: t.name,
        class: n("test-spec-name", [it])
      }, [s("div", {
        class: n("test-spec-name-container", [nl])
      }, [e.status === "complete" ? c.fail > 0 ? s(_e, {
        class: n("test-spec-complete", [$e, an])
      }, {
        default: () => [w("Fail")]
      }) : s(_e, {
        class: n("test-spec-pass", [$e, $o])
      }, {
        default: () => [w("Pass")]
      }) : s(_e, {
        class: n("test-spec-run", [$e, qo])
      }, {
        default: () => [w("Run")]
      }), s(cn, {
        onClick: () => {
          e.hideTestsAndSupressLogs || e.openSpec(t.name);
        },
        path: t.name
      }, null)]), e.verbose && !e.hideTestsAndSupressLogs && s(Hn, {
        tests: o
      }, null), e.verbose && !e.hideTestsAndSupressLogs && s(Un, {
        describes: l
      }, null), !e.hideTestsAndSupressLogs && Zo(t).map((d) => s("div", {
        key: `failing-${d.name}`,
        class: n("test-spec-error", [it])
      }, [s("div", {
        class: n("test-spec-error-text", [sl, ue])
      }, [w(" "), d.blocks.join("  "), w("  "), d.name]), d.errors.map((a) => s(rn, {
        key: `failing-${d.name}-error`,
        error: a,
        path: d.path
      }, null))]))]);
    })]);
  }
}), _e = A({
  name: "SpecLabel",
  setup(e, { slots: n, attrs: t }) {
    const o = R();
    return () => s("span", {
      class: o("test-spec-label", [il, (t == null ? void 0 : t.class) || ""])
    }, [n.default ? n.default() : null]);
  }
}), cn = A({
  name: "FilePath",
  props: {
    onClick: {
      type: Function,
      required: !0
    },
    path: {
      type: String,
      required: !0
    }
  },
  setup(e) {
    const n = k(() => e.path.split("/")), t = k(() => n.value.slice(0, n.value.length - 1).join("/") + "/"), o = k(() => n.value[n.value.length - 1]), l = R();
    return () => s("button", {
      class: l("test-filename", [le, ol]),
      onClick: () => {
        e.onClick();
      },
      type: "button"
    }, [s("span", {
      class: l("test-filename-base", [ll])
    }, [t.value]), s("span", {
      class: l("test-filename-file", [al])
    }, [o.value])]);
  }
}), un = g, ot = g, cl = g, ul = A({
  name: "Summary",
  props: {
    suites: {
      type: Object,
      required: !0
    },
    tests: {
      type: Object,
      required: !0
    },
    duration: {
      type: Number,
      required: !0
    }
  },
  setup(e) {
    const n = R(), t = "Test suites: ", o = (l) => {
      const c = t.length - l.length, d = Array.from({
        length: c
      }, () => " ").join("");
      return l + d;
    };
    return () => s("div", {
      class: n("test-summary", [cl])
    }, [s("div", {
      class: n("test-summary", [un])
    }, [s("span", {
      class: n("test-summary-suites-label", [ot])
    }, [t]), e.suites.fail > 0 && s("span", {
      class: n("test-summary-suites-fail", [ue])
    }, [e.suites.fail, w(" failed,"), " "]), e.suites.pass > 0 && s("span", {
      class: n("test-summary-suites-pass", [we])
    }, [e.suites.pass, w(" passed,"), " "]), s("span", null, [e.suites.total, w(" total")])]), s("div", {
      class: n("test-summary", [un])
    }, [s("span", {
      class: n("test-summary-label", [ot])
    }, [o("Tests:")]), e.tests.fail > 0 && s("span", {
      class: n("test-summary-fail", [ue])
    }, [e.tests.fail, w(" failed,"), " "]), e.tests.skip > 0 && s("span", {
      class: n("test-summary-skip", [_n])
    }, [e.tests.skip, w(" skipped,"), " "]), e.tests.pass > 0 && s("span", {
      class: n("test-summary-pass", [we])
    }, [e.tests.pass, w(" passed,"), " "]), s("span", null, [e.tests.total, w(" total")])]), s("div", {
      class: n("test-summary-curation", [ot])
    }, [o("Time:"), e.duration / 1e3, w("s")])]);
  }
}), dl = g, fl = {
  specs: {},
  status: "initialising",
  verbose: !1,
  watchMode: !0,
  suiteOnly: !1,
  specsCount: 0
}, pl = A({
  name: "SandpackTests",
  props: {
    verbose: {
      type: Boolean,
      required: !1,
      default: !1
    },
    watchMode: {
      type: Boolean,
      required: !1,
      default: !0
    },
    onComplete: {
      type: Function,
      required: !1,
      default: void 0
    },
    actionsChildren: {
      type: Object,
      required: !1,
      default: null
    },
    showVerboseButton: {
      type: Boolean,
      required: !1,
      default: !0
    },
    showWatchButton: {
      type: Boolean,
      required: !1,
      default: !0
    },
    hideTestsAndSupressLogs: {
      type: Boolean,
      required: !1,
      default: !1
    },
    standalone: {
      type: Boolean,
      required: !1,
      default: !1
    }
  },
  setup(e, { slots: n, attrs: t }) {
    let o, l;
    const c = /.*\.(test|spec)\.[tj]sx?$/, d = Sn(), { getClient: a, iframe: f, listen: y, sandpack: m } = vt(), h = R(), u = T(O(x({}, fl), {
      verbose: e.verbose || !1,
      watchMode: e.watchMode || !0
    })), C = () => {
      u.value.status = "running", u.value.specs = {};
      const F = a();
      F && F.dispatch({
        type: "run-all-tests"
      });
    }, r = () => {
      u.value = O(x({}, u.value), {
        status: "running",
        specs: {}
      });
      const F = a();
      F && F.dispatch({
        type: "run-tests",
        path: m.activeFile
      });
    }, P = k(() => m.activeFile.match(c) !== null), D = () => {
      let F = [], S = "";
      l && l(), l = y((v) => {
        if (!(u.value.suiteOnly && ("path" in v && v.path !== m.activeFile || "test" in v && "path" in v.test && v.test.path !== m.activeFile))) {
          if (v.type === "action" && v.action === "clear-errors" && v.source === "jest") {
            S = v.path;
            return;
          }
          if (v.type === "test") {
            if (v.event === "initialize_tests") {
              if (F = [], S = "", u.value.watchMode)
                return C();
              u.value = O(x({}, u.value), {
                status: "idle",
                specs: {}
              });
              return;
            }
            if (v.event === "test_count") {
              u.value = O(x({}, u.value), {
                specsCount: v.count
              });
              return;
            }
            if (v.event === "total_test_start") {
              F = [], u.value = O(x({}, u.value), {
                status: "running"
              });
              return;
            }
            if (v.event === "total_test_end") {
              e.onComplete !== void 0 && e.onComplete(u.value.specs), u.value = O(x({}, u.value), {
                status: "complete"
              });
              return;
            }
            if (v.event === "add_file") {
              u.value = pe(u.value, ["specs", v.path], {
                describes: {},
                tests: {},
                name: v.path
              });
              return;
            }
            if (v.event === "remove_file") {
              const E = Object.entries(u.value.specs).reduce((M, [j, $]) => j === v.path ? M : O(x({}, M), {
                [j]: $
              }), {});
              u.value = O(x({}, u.value), {
                specs: E
              });
              return;
            }
            if (v.event === "file_error") {
              u.value = pe(u.value, ["specs", v.path, "error"], v.error);
              return;
            }
            if (v.event === "describe_start") {
              F.push(v.blockName);
              const [E, M] = Pe(F), j = S;
              if (M === void 0)
                return;
              u.value = pe(u.value, ["specs", j, "describes", ...Me(E, ($) => [$, "describes"]), M], {
                name: v.blockName,
                tests: {},
                describes: {}
              });
              return;
            }
            if (v.event === "describe_end") {
              F.pop();
              return;
            }
            if (v.event === "add_test") {
              const [E, M] = Pe(F), j = {
                status: "idle",
                errors: [],
                name: v.testName,
                blocks: [...F],
                path: v.path
              };
              if (M === void 0) {
                u.value = pe(u.value, ["specs", v.path, "tests", v.testName], j);
                return;
              } else {
                u.value = pe(u.value, ["specs", v.path, "describes", ...Me(E, ($) => [$, "describes"]), M, "tests", v.testName], j);
                return;
              }
            }
            if (v.event === "test_start") {
              const { test: E } = v, [M, j] = Pe(E.blocks), $ = {
                status: "running",
                name: E.name,
                blocks: E.blocks,
                path: E.path,
                errors: []
              };
              if (j === void 0) {
                u.value = pe(u.value, ["specs", E.path, "tests", E.name], $);
                return;
              } else {
                u.value = pe(u.value, ["specs", E.path, "describes", ...Me(M, (ee) => [ee, "describes"]), j, "tests", E.name], $);
                return;
              }
            }
            if (v.event === "test_end") {
              const { test: E } = v, [M, j] = Pe(E.blocks), $ = {
                status: E.status,
                errors: E.errors,
                duration: E.duration,
                name: E.name,
                blocks: E.blocks,
                path: E.path
              };
              j === void 0 ? u.value = pe(u.value, ["specs", E.path, "tests", E.name], $) : u.value = pe(u.value, ["specs", E.path, "describes", ...Me(M, (ee) => [ee, "describes"]), j, "tests", E.name], $);
            }
          }
        }
      });
    }, J = () => {
      o && o(), o = y(({ type: F }) => {
        F === "done" && u.value.watchMode && (P.value ? r() : C());
      });
    };
    Rt(J), Rt(D), xe(J), ae(() => {
      o && o(), l && l();
    }), se(() => {
      o && o(), l && l();
    });
    const W = (F) => {
      m.setActiveFile(F);
    }, L = k(() => Object.values(u.value.specs)), I = k(() => Ko(L.value)), K = k(() => Wo(L.value)), re = k(() => Go(L.value));
    return () => s(Ee, me(e, {
      class: h("tests", [(t == null ? void 0 : t.class) || ""]),
      style: x(x({}, Bo(d.mode === "dark")), (t == null ? void 0 : t.style) || {})
    }), {
      default: () => [s("iframe", {
        ref: f,
        style: {
          display: "none"
        },
        title: "Sandpack Tests"
      }, null), s(Ho, {
        hideTestsAndSupressLogs: e.hideTestsAndSupressLogs,
        showVerboseButton: e.showVerboseButton,
        showWatchButton: e.showWatchButton,
        setSuiteOnly: () => {
          u.value.suiteOnly = !u.value.suiteOnly;
        },
        setVerbose: () => {
          u.value = O(x({}, u.value), {
            verbose: !u.value.verbose
          });
        },
        setWatchMode: () => {
          u.value = O(x({}, u.value), {
            watchMode: !u.value.watchMode
          });
        },
        showSuitesOnly: u.value.specsCount > 1,
        status: u.value.status,
        suiteOnly: u.value.suiteOnly,
        verbose: u.value.verbose,
        watchMode: u.value.watchMode
      }, null), u.value.status === "running" || u.value.status === "initialising" ? s(Mn, {
        showOpenInCodeSandbox: !1
      }, null) : s("div", {
        class: dl.toString()
      }, [n.actionsChildren ? n.actionsChildren() : e.actionsChildren ? e.actionsChildren : null, s(Vo, {
        onClick: () => {
          u.value.suiteOnly ? r() : C();
        }
      }, null)]), s("div", {
        class: ml.toString()
      }, [L.value.length === 0 && u.value.status === "complete" ? s("div", {
        class: vl.toString()
      }, [s("p", null, [w("No test files found.")]), s("p", null, [w("Test match:"), " ", s("span", {
        class: ue.toString()
      }, [c.toString()])])]) : s(Q, null, [s(rl, {
        openSpec: W,
        specs: L.value,
        status: u.value.status,
        verbose: u.value.verbose,
        hideTestsAndSupressLogs: e.hideTestsAndSupressLogs
      }, null), u.value.status === "complete" && K.value.total > 0 && s(ul, {
        duration: I.value,
        suites: re.value,
        tests: K.value
      }, null)])])]
    });
  }
}), ml = g, vl = g, hl = A({
  name: "SandpackRender",
  props: {
    fragment: {
      type: Boolean,
      required: !1,
      default: !1
    }
  },
  setup(e, { slots: n }) {
    return () => e.fragment ? n.default ? n.default() : null : s(Ee, null, {
      default: () => [n.default ? n.default() : null]
    });
  }
}), yl = {
  files: {
    type: Object,
    required: !1,
    default: void 0
  },
  template: {
    type: String,
    required: !1,
    default: void 0
  },
  customSetup: {
    type: Object,
    required: !1,
    default: void 0
  },
  theme: {
    type: [String, Object],
    required: !1,
    default: void 0
  },
  options: {
    type: Object,
    required: !1,
    default: void 0
  },
  rtl: {
    type: Boolean,
    required: !1,
    default: !1
  }
}, gl = A({
  name: "Sandpack",
  props: yl,
  setup(e) {
    var I, K, re, F;
    const n = T(0), t = T(null), o = T((K = (I = e.options) == null ? void 0 : I.editorWidthPercentage) != null ? K : 50), l = T(70), c = T((F = (re = e.options) == null ? void 0 : re.showConsole) != null ? F : !1), d = k(() => {
      var S, v;
      return (v = (e == null ? void 0 : e.rtl) || ((S = e == null ? void 0 : e.options) == null ? void 0 : S.rtl)) != null ? v : !1;
    }), a = k(() => {
      var S, v;
      return (v = (S = e.options) == null ? void 0 : S.resizablePanels) != null ? v : !0;
    }), f = k(() => {
      var S, v, E, M, j, $, ee, ie, ce, oe, de, ve, he, ye;
      return {
        showTabs: (S = e.options) == null ? void 0 : S.showTabs,
        showLineNumbers: (v = e.options) == null ? void 0 : v.showLineNumbers,
        showInlineErrors: (E = e.options) == null ? void 0 : E.showInlineErrors,
        wrapContent: (M = e.options) == null ? void 0 : M.wrapContent,
        closableTabs: (j = e.options) == null ? void 0 : j.closableTabs,
        initMode: ($ = e.options) == null ? void 0 : $.initMode,
        extensions: (ie = (ee = e.options) == null ? void 0 : ee.codeEditor) == null ? void 0 : ie.extensions,
        extensionsKeymap: (oe = (ce = e.options) == null ? void 0 : ce.codeEditor) == null ? void 0 : oe.extensionsKeymap,
        readOnly: (de = e.options) == null ? void 0 : de.readOnly,
        showReadOnly: (ve = e.options) == null ? void 0 : ve.showReadOnly,
        additionalLanguages: (ye = (he = e.options) == null ? void 0 : he.codeEditor) == null ? void 0 : ye.additionalLanguages
      };
    }), y = k(() => {
      var S, v, E, M, j, $, ee, ie, ce, oe, de, ve, he, ye, Ne;
      return {
        activeFile: (S = e.options) == null ? void 0 : S.activeFile,
        visibleFiles: (v = e.options) == null ? void 0 : v.visibleFiles,
        recompileMode: (E = e.options) == null ? void 0 : E.recompileMode,
        recompileDelay: (M = e.options) == null ? void 0 : M.recompileDelay,
        autorun: (j = e.options) == null ? void 0 : j.autorun,
        autoReload: ($ = e.options) == null ? void 0 : $.autoReload,
        bundlerURL: (ee = e.options) == null ? void 0 : ee.bundlerURL,
        startRoute: (ie = e.options) == null ? void 0 : ie.startRoute,
        skipEval: (ce = e.options) == null ? void 0 : ce.skipEval,
        fileResolver: (oe = e.options) == null ? void 0 : oe.fileResolver,
        initMode: (de = e.options) == null ? void 0 : de.initMode,
        initModeObserverOptions: (ve = e.options) == null ? void 0 : ve.initModeObserverOptions,
        externalResources: (he = e.options) == null ? void 0 : he.externalResources,
        logLevel: (ye = e.options) == null ? void 0 : ye.logLevel,
        classes: (Ne = e.options) == null ? void 0 : Ne.classes
      };
    }), m = k(() => {
      var S, v;
      return {
        flexGrow: 100 - o.value,
        flexShrink: 100 - o.value,
        flexBasis: 0,
        width: 100 - o.value + "%",
        gap: c.value ? "1px" : "0",
        height: (S = e.options) != null && S.editorHeight ? `${(v = e.options) == null ? void 0 : v.editorHeight}px` : void 0
      };
    }), h = k(() => {
      var S;
      return (S = ct[e.template]) != null ? S : {};
    }), u = k(() => {
      var S, v;
      return (S = e.options) != null && S.layout ? (v = e.options) == null ? void 0 : v.layout : "mode" in h.value ? h.value.mode : "preview";
    }), C = k(() => {
      var S;
      return (S = e.options) != null && S.showConsoleButton ? s(bl, {
        counter: n.value,
        onClick: () => {
          c.value = !c.value;
        }
      }, null) : void 0;
    }), r = k(() => {
      var S, v;
      return ((S = e.options) == null ? void 0 : S.showConsole) || ((v = e.options) == null ? void 0 : v.showConsoleButton);
    }), P = k(() => r.value ? {
      class: U + "-preset-column",
      style: m.value
    } : {}), D = k(() => r.value ? {
      flexGrow: l.value,
      flexShrink: l.value,
      flexBasis: 0,
      overflow: "hidden"
    } : m.value), J = (S) => {
      if (!t.value)
        return;
      S.preventDefault();
      const v = t.value.parentElement;
      if (!v)
        return;
      const M = t.value.dataset.direction === "horizontal", { left: j, top: $, height: ee, width: ie } = v.getBoundingClientRect(), ce = M ? (S.clientX - j) / ie * 100 : (S.clientY - $) / ee * 100, oe = Math.min(Math.max(ce, 25), 75);
      M ? o.value = d.value ? 100 - oe : oe : l.value = oe, v.querySelectorAll(`.${U}-stack`).forEach((de) => {
        de.style.pointerEvents = "none";
      });
    }, W = () => {
      var v;
      const S = (v = t.value) == null ? void 0 : v.parentElement;
      S && (S.querySelectorAll(`.${U}-stack`).forEach((E) => {
        E.style.pointerEvents = "";
      }), t.value = null);
    }, L = R();
    return H([() => {
      var S;
      return (S = e.options) == null ? void 0 : S.showConsole;
    }], () => {
      var S, v;
      c.value = (v = (S = e.options) == null ? void 0 : S.showConsole) != null ? v : !1;
    }, {
      immediate: !0
    }), H([() => e.options], () => {
      a.value && typeof document < "u" && (document.body.removeEventListener("mousemove", J), document.body.removeEventListener("mouseup", W), document.body.addEventListener("mousemove", J), document.body.addEventListener("mouseup", W));
    }, {
      immediate: !0
    }), ae(() => {
      typeof document < "u" && (document.body.removeEventListener("mousemove", J), document.body.removeEventListener("mouseup", W));
    }), () => s(gn, me(e, {
      key: e.template,
      customSetup: e.customSetup,
      files: e.files,
      options: y.value,
      template: e.template,
      theme: e.theme
    }), {
      default: () => [s(Nn, {
        class: d.value ? L("rtl-layout", [Sl]) : ""
      }, {
        default: () => {
          var S;
          return [s(In, me(f.value, {
            style: {
              height: `${(S = e.options) == null ? void 0 : S.editorHeight}px`,
              flexGrow: o.value,
              flexShrink: o.value,
              flexBasis: 0,
              overflow: "hidden"
            }
          }), null), a.value && s("div", {
            class: L("resize-handler", [dn()]),
            "data-direction": "horizontal",
            onMousedown: (v) => {
              t.value = v.target;
            },
            style: {
              left: `calc(${d.value ? 100 - o.value : o.value}% - 5px)`
            }
          }, null), s(hl, me({
            fragment: !r.value
          }, P.value), {
            default: () => {
              var v, E, M;
              return [u.value === "preview" && s(qn, {
                actionsChildren: C.value,
                showNavigator: (v = e.options) == null ? void 0 : v.showNavigator,
                showRefreshButton: (E = e.options) == null ? void 0 : E.showRefreshButton,
                style: D.value
              }, null), u.value === "tests" && s(pl, {
                standalone: !0,
                actionsChildren: C.value,
                style: D.value
              }, null), u.value === "console" && s(ln, {
                actionsChildren: C.value,
                style: D.value
              }, null), (((M = e.options) == null ? void 0 : M.showConsoleButton) || c.value) && s(Q, null, [a.value && c.value && s("div", {
                class: L("resize-handler", [dn()]),
                "data-direction": "vertical",
                onMousedown: (j) => {
                  t.value = j.target;
                },
                style: {
                  top: `calc(${l.value}% - 5px)`
                }
              }, null), s("div", {
                class: L("console-wrapper", [wl]),
                style: {
                  flexGrow: c.value ? 100 - l.value : 0,
                  flexShrink: c.value ? 100 - l.value : 0,
                  flexBasis: 0
                }
              }, [s(ln, {
                onLogsChange: (j) => {
                  n.value = j.length;
                },
                showHeader: !1
              }, null)])])];
            }
          })];
        }
      })]
    });
  }
}), bl = A({
  props: {
    onClick: {
      type: Function,
      required: !0
    },
    counter: {
      type: Number,
      required: !0
    }
  },
  setup(e) {
    return () => s(Se, {
      class: Cl.toString(),
      onClick: e.onClick
    }, {
      default: () => [s(ht, null, null), e.counter > 0 && s("strong", null, [e.counter])]
    });
  }
}), dn = g, Cl = g, wl = g, Sl = g;
function Vl() {
  return {
    install(e) {
      e.component("Sandpack", gl), e.component("SandpackLayout", Nn), e.component("SandpackProvider", gn), e.component("ErrorOverlay", gt), e.component("LoadingOverlay", bt), e.component("CodeEditor", Ze), e.component("SandpackCodeEditor", In), e.component("SandpackCodeViewer", Pn), e.component("SandpackPreview", qn), e.component("SandpackStack", Ee), e.component("SandpackTranspiledCode", xo), e.component("SandpackFileExplorer", mo);
    }
  };
}
export {
  ci as ANGULAR_TEMPLATE,
  Oi as BackwardIcon,
  Fi as CleanIcon,
  Bi as CloseIcon,
  Ze as CodeEditor,
  ht as ConsoleIcon,
  ao as DependenciesProgress,
  fo as Directory,
  Ii as DirectoryIconClosed,
  Ri as DirectoryIconOpen,
  gt as ErrorOverlay,
  ji as ExportIcon,
  Cn as FADE_ANIMATION_DURATION,
  Dn as File,
  Di as FileIcon,
  An as FileTabs,
  Mi as ForwardIcon,
  bt as LoadingOverlay,
  Bn as ModuleList,
  go as Navigator,
  On as OpenInCodeSandboxButton,
  ui as REACT_TEMPLATE,
  di as REACT_TYPESCRIPT_TEMPLATE,
  Tn as RefreshIcon,
  We as RestartIcon,
  Se as RoundedButton,
  Ln as RunButton,
  kn as RunIcon,
  ct as SANDBOX_TEMPLATES,
  Ds as SANDPACK_THEMES,
  fi as SOLID_TEMPLATE,
  pi as SVELTE_TEMPLATE,
  gl as Sandpack,
  In as SandpackCodeEditor,
  Pn as SandpackCodeViewer,
  ln as SandpackConsole,
  mo as SandpackFileExplorer,
  Nn as SandpackLayout,
  qn as SandpackPreview,
  gn as SandpackProvider,
  Ee as SandpackStack,
  yn as SandpackStateContext,
  pl as SandpackTests,
  wn as SandpackThemeContext,
  ki as SandpackThemeProvider,
  xo as SandpackTranspiledCode,
  Pi as SignInIcon,
  qi as SignOutIcon,
  mi as TEST_TYPESCRIPT_TEMPLATE,
  Gi as UnstyledOpenInCodeSandboxButton,
  vi as VANILLA_TEMPLATE,
  hi as VANILLA_TYPESCRIPT_TEMPLATE,
  si as VITE_PREACT_TEMPLATE,
  ii as VITE_PREACT_TS_TEMPLATE,
  ti as VITE_REACT_TEMPLATE,
  ni as VITE_REACT_TS_TEMPLATE,
  oi as VITE_SVELTE_TEMPLATE,
  li as VITE_SVELTE_TS_TEMPLATE,
  ei as VITE_TEMPLATE,
  ai as VITE_VUE_TEMPLATE,
  ri as VITE_VUE_TS_TEMPLATE,
  yi as VUE2_TEMPLATE,
  gi as VUE2_TS_TEMPLATE,
  zt as VUE_TEMPLATE,
  Jt as VUE_TS_TEMPLATE,
  Hl as createContext,
  Vl as default,
  at as defaultDark,
  Te as defaultLight,
  _l as getSandpackCssText,
  zi as layoutClassName,
  yt as stackClassName,
  Vi as tabButton,
  bn as useActiveCode,
  R as useClassNames,
  mt as useContext,
  Si as useLoadingOverlayState,
  z as useSandpack,
  vt as useSandpackClient,
  Do as useSandpackConsole,
  xi as useSandpackNavigation,
  Ye as useSandpackShell,
  Sn as useSandpackTheme
};
